# METHODS {#METHODS}

\vspace*{-18pt}

{{< include /_setup.qmd >}}

```{r MethodsVariables, echo=FALSE}

methods_variables <- list()  # Initialize list for assigned variables

# Calculate SEINet records with non-missing geographic coordinates.
methods_variables$seinet <-
  dplyr::filter(seinet_coords, !is.na(Latitude), !is.na(Longitude))

# Read in raw FASTA system files.
methods_variables$fasta_paths <-
  list.files(
    path = system.file("extdata/FASTA", package = "thesis"),
    full.names = TRUE
  )

# Create list of Biostrings objects from FASTA files.
methods_variables$fasta_msa <-
  purrr::map(.x = methods_variables$fasta_paths, function(fasta) {
    Biostrings::readDNAStringSet(filepath = fasta)
  })

# Add names to raw fasta files
names(methods_variables$fasta_msa) <-
  gsub(pattern = "-combined_raw.fasta", replacement = "",
       x = fs::path_file(methods_variables$fasta_paths))

```

## Herbarium Specimens

\paragraph{} Specimens of *Physaria* sensu stricto were requested from the
following herbaria:
RM, NY, GH, MONT, MONTU, IDS, F, UC, RSA-POM, MO, UTC, and US. In total,
`r count_specimens(spp_tibble = herbarium_specimens)` unique collections were
reviewed. For each specimen voucher, the identification history, collector,
collection number and date, institution, locality, geographic coordinates,
elevation, and ecological description were recorded.
Measurements and observations of continuous and discrete character traits were
collected for morphological analyses and phenotypic descriptions.
Montana State University’s Graphical Locator for Township, Range, and Section
(TRS) data [@Wefald2003] was used to approximate collection locations to the
center of a section. When coordinate data were not provided, Google Earth
[@GoogleEarth] was used to estimate coordinates to half-minute precision
based on the description of collection location. Specimens with ambiguous
locations (e.g. a river or canyon) were omitted from distribution mapping.
`r R.version$version.string` [@R-base] was used to read data from an *.xlsx*
file with the `readxl` package [@R-readxl], filter annotation records to those
containing either  *Physaria* or *Lesquerella* with the `dplyr` package
[@R-dplyr], and standardize collection dates into the date format "MM/DD/YYYY"
using the `lubridate` package [@R-lubridate; @R-lubridate2011].
Prior annotations were parsed using the `purrr` [@R-purrr] and `stringr`
packages [@R-stringr] to substitute the species name for identification
agreements and replace identification synonyms following @OKane2010.
Elevation data were converted from meters to feet for unit consistency.

\paragraph{} Specimen distribution maps were established using location data
collected from herbarium voucher labels as geographic coordinates converted
into decimal degree format. Administrative boundaries (state and county
borders) were obtained from the United States Census Bureau TIGER/Line
shapefile database through the `tigris` package [@R-tigris].
Raster images for satellite maps and digital elevation model data were obtained
by the `ggmap` [@R-ggmap] and `elevatr` [@R-elevatr] packages, respectively.
Specimen data were mapped using `ggplot2` [@R-ggplot2], with coordinate data
treated as simple features by the `sf` package [@R-sf].
Additional specimen occurrence data were obtained from the
Southwest Environmental Information Network (SEINet) data portal [@SEINet].
Two specimen searches of all available herbaria were performed using the
scientific name as taxonomic criteria with synonyms included for
"Physaria floribunda" and "Physaria bellii" as search terms.
Occurrence records were downloaded as *.csv* files in the Darwin Core structure
[@Wieczorek2012]. Searches returned a total of `r nrow(seinet_coords)` records,
`r nrow(methods_variables$seinet)` with specimen occurrence
decimal degree coordinates.

\paragraph{} This mapping framework was used to investigate the biogeographic
distribution of continuous and discrete taxonomically-informative morphological
traits of all considered *Physaria* species for replum shape, ovules per locule,
mature fruit apices and trichomes, and basal leaf margins.
Additional assessment of morphological differences included maximum fruit size
for the putative Carbon County, Wyoming taxon, and basal leaf and stem lengths
for the three condensed growth habit species *P. dornii*, *P. condensata*, and
*P. eburniflora* in the context of their suspected sister taxa. For the putative
Carbon County taxon, elevation differences were assessed by one-way ANOVA with
species as the independent variable, and fruit length by analysis of covariance
treating collection date and species as independent variables. Functions for
handling specimen trait data and plotting distribution maps, morphology, and
phylogenetic trees made use of additional R Tidyverse packages:
`tibble` [@R-tibble], `plyr` [@R-plyr], `tidyr` [@R-tidyr], `readr` [@R-readr],
`magrittr` [@R-magrittr], `rlang` [@R-rlang], `fs` [@R-fs], and
`broom` [@R-broom]. Tables were generated using `knitr` [@R-knitr]
and formatted with `kableExtra` [@R-kableExtra]. Plot figures were formatted
using `cowplot` [@R-cowplot], `ggtext` [@R-ggtext], `scales` [@R-scales], and
`ggnewscale` [@R-ggnewscale].

## Tissue Sources

\paragraph{} In the context of molecular phylogenies, there has been debate
regarding the extent to which density of taxon sampling versus gene number
affect the accuracy of phylogenetic inference. Simulation studies by
@MaddisonKnowles2006 found for recent divergences, sampling more individuals
increased the accuracy of species tree inference. In such cases, increased taxon
sampling may improve the number of detectable gene coalescence events between
sister taxa. In contrast, simulations by @RokasCarroll2005 found increasing the
number of genetic loci analyzed improved the accuracy of analysis while taxon
sampling had either no statistically significant effect or a slight negative
correlation. Therefore, samples were chosen from across each species’ range to
broadly represent populations within their distribution and multiple loci were
sampled for each individual.

\paragraph{} Taxon sampling was determined by species distribution patterns,
with broader sampling from more widespread species (Table
`r knitr_table_dna(knit_variables$pandoc)`). For the narrowly distributed
Wyoming state endemics *P. dornii* and *P. condensata*, four individuals were
sampled from each taxa. For Wyoming state endemics *P. eburniflora* and
*P. saximontana* subsp. *saximontana*, five individuals from each taxa were
included. Four individuals of *P. saximontana* subsp. *dentata* were sampled
broadly from Montana. Five specimens identified as *P. didymocarpa* subsp.
*lanata* in Wyoming and three identified as *P. didymocarpa* subsp. *lyrata* in
Idaho were included. These two *P. didymocarpa* subspecies have narrow
distributions and are thought to be state endemics of Wyoming and Idaho,
respectively. Five individuals were sampled from *P. didymocarpa* subsp.
*didymocarpa* in Wyoming and Montana. For the regional endemics
*P. brassicoides* and *P. integrifolia*, six and five individuals were sampled,
respectively. For *P. vitulifera*, three individuals from Colorado and four
putative specimens with annotations as *P. vitulifera* from Carbon County,
Wyoming were included. The Carbon County population is of interest due to the
questionable status of this species’ distribution in the state of Wyoming
[@HeidelOKane2012]. Finally, twelve individuals of the widespread
*P. acutifolia* from Utah, Colorado, and Wyoming were sampled.
Sequence data were obtained from
`r length(methods_variables$fasta_msa$rITS)` specimens for the rITS locus,
`r length(methods_variables$fasta_msa$rps)` for the *rps*
intron, and `r length(methods_variables$fasta_msa$ycf1)` for *ycf1*.
Analyses of concatenated multi-locus or partitioned data were limited to
specimens for which sequencing of all three loci was completed.

## DNA Extraction

\paragraph{} Genomic DNA was extracted from one of two sources: basal leaf
tissue taken from loaned herbarium specimens (ca. 20 mg) or silica bead-dried
leaf tissue from specimens collected in June and August of 2014 from Wyoming,
Idaho, and Montana with O’Kane Jr. Dried leaf tissues were stored at
-20 $^{\circ}$C prior to extraction. DNA was extracted from a single
individual specimen to minimize noise in the sequence data. Sequence data from
a preliminary systematic analysis of *Physaria* sensu stricto by
@Rethwisch2014 in addition to data previously obtained by O'Kane were
also included in this study.

\clearpage

```{r TableDnaSpecimens}

# Read in DNA collection localities for methods table.
methods_variables$collections <-
  list.files(system.file("extdata/Methods", package = "thesis"),
             pattern = "DNA-collections", full.names = TRUE) %>%
  readr::read_csv(file = ., col_names = TRUE, col_types = "ccc") %>%

  # Isolate surnames for collectors.
  dplyr::mutate(`Voucher Information` = sapply(
      `Voucher Information`, function(voucher) {
        gsub("[A-Z]\\. ?", "", voucher) %>% gsub("&", "and", x = .)
        }))

# Table footnote caption
methods_variables$dna_footnote <-
  paste(
    "Requested herbarium specimens used for extraction of",
    "genomic DNA. Voucher information includes record",
    "collector and collection number while locality reports",
    "the county and state for the collection. Specimens",
    "with a single * represent vouchers with DNA sequence",
    "data obtained by Rethwisch (2014) while a double **",
    "indicate specimens with sequence data previously obtained by",
    "O'Kane Jr."
  )

```

```{r TableDnaSpecimensLatex}

if (knit_variables$pandoc == "latex") {
  methods_variables$collections %>% dplyr::slice(1:39) %>%
    knitr::kable(x =.,
      format = knit_variables$pandoc, escape = FALSE,
      caption = "DNA extraction specimens.",
      caption.short = "DNA extraction specimens"
    ) %>%
    kableExtra::kable_styling(
      full_width = FALSE, font_size = 10,
      latex_options= "hold_position"
    ) %>%
    kableExtra::row_spec(row = 0, bold = TRUE, font_size = 10) %>%
    kableExtra::column_spec(1, border_left = TRUE, width = "3.2cm") %>%
    kableExtra::column_spec(2, width = "6cm") %>%
    kableExtra::column_spec(3, border_right = TRUE, width = "3.5cm") %>%
    kableExtra::collapse_rows(columns = 1) %>%
    kableExtra::footnote(
      general = methods_variables$dna_footnote,
      general_title = "",
      threeparttable = TRUE,
      escape = FALSE
    )
}

```

\clearpage

<!--- Clear caption formatting to prevent automatic labelling -->
\captionsetup[table]{labelformat=empty}

<!--- Prevent automatic indexing of table entry in List of Tables -->
\captionsetup[table]{list=no}

```{r TableDnaSpecimensLatexContinued}

if (knit_variables$pandoc == "latex") {
  methods_variables$collections %>% dplyr::slice(40:68) %>%
    knitr::kable(x = .,
      format = knit_variables$pandoc, escape = FALSE,
      caption = "Table 2. Continued."
    ) %>%
    kableExtra::kable_styling(
      full_width = FALSE, font_size = 10,
      latex_options= "hold_position"
    ) %>%
    kableExtra::row_spec(row = 0, bold = TRUE, font_size = 10) %>%
    kableExtra::column_spec(1, border_left = TRUE, width = "3.2cm") %>%
    kableExtra::column_spec(2, width = "6cm") %>%
    kableExtra::column_spec(3, border_right = TRUE, width = "3.5cm") %>%
    kableExtra::collapse_rows(columns = 1)
}

```

<!--- Reset caption options and table counter -->
\captionsetup[table]{labelformat=default}
\captionsetup[table]{list=yes}
\setcounter{table}{2}

```{r TableDnaSpecimensHtml}

if (knit_variables$pandoc != "latex") {
  methods_variables$collections <- methods_variables$collections %>%
    dplyr::mutate(taxa = purrr::map_chr(.data$Taxon, function(taxon) {
      gsub("\\\\textit\\{", "\\*", taxon) %>%
        gsub("\\}$", "\\*", x = .) %>%
        gsub("\\} subsp.", "\\*  subsp.", x = .)
    })) %>%
    dplyr::select(c(taxa, `Voucher Information`, Locality)) %>%
    dplyr::rename(Taxon = taxa) %>%
    knitr::kable(
      longtable = TRUE, escape = FALSE,
      format = knit_variables$kable_format,
      caption = "DNA Extraction Herbarium Specimens"
    )
}

if (knit_variables$pandoc == "html") {
    methods_variables$collections %>%
    kableExtra::kable_styling(full_width = TRUE, font_size = 10) %>%
    kableExtra::row_spec(row = 0, bold = TRUE, font_size = 12) %>%
    kableExtra::column_spec(1, border_left = TRUE, bold = TRUE) %>%
    kableExtra::column_spec(2, width = "6cm") %>%
    kableExtra::column_spec(3, width = "3.5cm", border_right = TRUE) %>%
    kableExtra::collapse_rows(columns = 1) %>%
    kableExtra::footnote(
      general = methods_variables$dna_footnote,
      general_title = "",
      threeparttable = TRUE,
      escape = FALSE
    )
}

if (knit_variables$pandoc == "docx") {
  methods_variables$collections
}

```

\vspace*{10pt}

\paragraph{} Genomic DNA was purified using a Qiagen DNeasy Plant Mini Kit.
Prior to the extraction protocol, leaf tissue was prepared using one of three
methods. Firstly, approximately 20 mg dried plant tissue was combined with ca.
250 mg washed sea sand and ground to a fine powder using a mortar and pestle.
Secondly, a modified approach of @Alexander2007 was used where ca.
20 mg of plant tissue was combined in a 2.0 mL microcentrifuge tube with three
4.5 mm zinc plated BBs and two 2.3 mm stainless steel beads. Samples were
attached to a reciprocating saw and the saw was operated for 90s at medium until
leave tissue was pulverized. Finally, a Retsch MM400 Bead Mill was used in place
of the reciprocating saw with the same bead, BB and tissue combination. Plant
tissue was disrupted for 45s at 20 hertz resulting in a fine ground powder.

\paragraph{} Once the plant tissue was prepared, Buffer AP1 (400 $\mu$L) and
RNase A (4 $\mu$L) were added. Samples were vortexed and incubated for 30
minutes at 65 $^{\circ}$C. Sample microcentrifuge tubes were inverted after
five, 15, and 25 minutes during incubation. Following incubation, Buffer P3
(130 $\mu$L) was added. Samples were then hand mixed and incubated on ice for
five minutes. The resultant lysate was centrifuged for five minutes at 14,000
rpm. Lysate supernatant was transferred to a QIAshredder spin column and
centrifuged for two minutes at 14,000 rpm. Flow-through was then transferred
to a new 1.5 mL microcentrifuge tube and mixed by pipette with 1.5x volume
Buffer AW1. 650 $\mu$L of the resulting mixture was transferred to a DNeasy
mini spin column and centrifuged for one minute at 9,000 rpm. Flow-through was
discarded before adding the remaining mixture to the column, and centrifuging
again for one minute at 9,000 rpm.

\paragraph{} Spin columns were then placed in new 2.0 mL collection tubes.
Buffer AW2 (500 $\mu$L) was added and the tubes were centrifuged for one minute
at 10,000 rpm. Flow-through was discarded before adding additional Buffer AW2
(500 $\mu$L) to each column. Spin columns were centrifuged for two minutes at
14,000 rpm before transfer to new 1.5 mL centrifuge tubes. Preheated Buffer AE
(75 $\mu$L at 65 $^{\circ}$C) was added to elute DNA from the columns. Samples
were incubated for 10 minutes at room temperature before centrifuging for one
minute at 9,000 rpm. Additional preheated Buffer AE (75 $\mu$L at 65
$^{\circ}$C) was added to each column. Samples were incubated at room
temperature for ten minutes before centrifuging an additional minute at 9,000
rpm. Modifications to the standard Qiagen DNA extraction protocol followed
recommendations made by @Drabkova2002. Time for the initial incubation
period was extended from 10 to 30 minutes to ensure complete cellular lysis.
Additionally, the volume of AE buffer used to elute DNA from the spin column was
reduced to 75 $\mu$L from 100 $\mu$L. The incubation time for column elution was
also extended from five minutes to 10 minutes.

## rITS PCR

\paragraph{} All Polymerase Chain Reaction (PCR) was completed using a Biometra
TProfessional Basic Thermocycler Gradient with individual or 8-tube strip 0.25
mL PCR microcentrifuge tubes. Individual reactions for the rITS region had a
total volume of 25 $\mu$L consisting of 13.85 $\mu$L distilled H2O (ddH2O),
2.5 $\mu$L 10X Reaction Buffer (500 mM KCl; 100 mM Tris-HCl, pH 8.8;
17.5 mM MgCl2; 0.5% v/v Triton-X; 5% v/v DMSO), 2.5 $\mu$L rITS-F primer
(2 $\mu$M, Table \@ref(tab:TablePCRPrimers)), 2.5 $\mu$L rITS-R primer
(2 $\mu$M, Table \@ref(tab:TablePCRPrimers)), 2.5 $\mu$L 8 mM dNTP mix
(10 $\mu$L 100 mM dATP; 10 $\mu$L 100 mM dGTP; 10 $\mu$L 100 mM dCTP; 10 $\mu$L
100 mM dTTP; 5 $\mu$L 1 M Tris-HCl, pH 7.9; 455 $\mu$L ddH2O), 0.15 $\mu$L Taq
DNA Polymerase (5 U/$\mu$L) and 1.0 $\mu$L sample DNA. Lid temperature for
thermocycling was set to 99 $^{\circ}$C. An initial DNA denaturation period of
two minutes at 94 $^{\circ}$C was followed by 30 cycles of denaturation, primer
annealing, and DNA extension. Each cycle consisted of denaturation for 45s at
94 $^{\circ}$C, primer annealing for one minute at 60 $^{\circ}$C, and
extension for one minute at 72 $^{\circ}$C. After 30 cycles, a final extension
of five minutes at 72 $^{\circ}$C was followed by a 16 $^{\circ}$C hold.

\vspace*{20pt}

```{r TablePCRPrimers}

methods_variables$primers <-
  system.file("extdata/Methods", package = "thesis") %>%
  list.files(
    pattern = "DNA-primers",
    full.names = TRUE
  ) %>%
  readr::read_csv(file = ., col_names = TRUE, col_types = "ccc")

if (knit_variables$pandoc != "latex") {
  methods_variables$primers <- methods_variables$primers %>%
    dplyr::mutate(
      Primer = stringr::str_replace_all(
        string = .data$Primer,
        pattern = "^\\\\textit\\{|\\}",
        replacement = "*"
      )
    )
}

methods_variables$primers <- methods_variables$primers %>%
  knitr::kable(x = .,
    format = knit_variables$kable_format,
    align = "l", escape = FALSE,
    caption = "PCR reaction primer sequences.",
    caption.short = "PCR reaction primer sequences"
  )

if (knit_variables$pandoc != "docx") {
  methods_variables$primers %>%
    kableExtra::kable_styling(
      full_width = FALSE,
      latex_options= "hold_position",
      position = "center"
    ) %>%
    kableExtra::row_spec(row = 0, bold = TRUE, font_size = 10) %>%
    kableExtra::column_spec(1, border_left = TRUE) %>%
    kableExtra::column_spec(3, border_right = TRUE)
} else {
  methods_variables$primers
}

```

\vspace*{10pt}

## *rps* Intron PCR

\paragraph{} For the *rps* intron, individual reactions had a total volume of
25 $\mu$L consisting of 11.25 $\mu$L ddH2O, 5.0 $\mu$L 5x Phusion HF Buffer
(proprietary, Thermo Fisher Scientific), 2.5 $\mu$L *rps*-F (2 $\mu$M, Table
(Table \@ref(tab:TablePCRPrimers))), 2.5 $\mu$L *rps*-R2 (2 $\mu$M,
Table (Table \@ref(tab:TablePCRPrimers))), 2.5 $\mu$L dNTPs, 0.25
$\mu$L Phusion Hot Start II High-Fidelity DNA Polymerase (Thermo Fisher
Scientific), and 1.0 $\mu$L sample DNA. Lid temperature for thermocycling was
set to 99 $^{\circ}$C. An initial 1 minute denaturation period at
98 $^{\circ}$C was followed by 30 cycles. Each cycle consisted of denaturation
for 10s at 98 $^{\circ}$C, annealing for 25s at 66.8 $^{\circ}$C, and extension
for 30s at 72 $^{\circ}$C. A final extension for seven minutes at 72
$^{\circ}$C was followed by a 16 $^{\circ}$C hold.

## *ycf1* PCR

\paragraph{} The total volume for *ycf1* reactions was 20 $\mu$L. Due to
inconsistencies in the success of PCR amplification, the volume of sample DNA
in the PCR reactions ranged from 1.0 $\mu$L to 2.0 $\mu$L. An equivalent
decrease in the volume of ddH2O, from 8.8 $\mu$L to 7.8 $\mu$L was used to
maintain a total 20 $\mu$L volume. In addition, the reaction mixture included
4.0 $\mu$L Phusion HF 5x buffer, 2.0 $\mu$L dNTPs, 2.0 $\mu$L *ycf1*-F
(2 $\mu$M, Table (Table \@ref(tab:TablePCRPrimers))), 2.0 $\mu$L
*ycf1*-R (2 $\mu$M, Table \@ref(tab:TablePCRPrimers)), and
0.2 $\mu$L Phusion DNA polymerase. Lid temperature for thermocycling was set
to 99 $^{\circ}$C. An initial denaturation period for one minute at 98
$^{\circ}$C was followed by 35 cycles. Each cycle consisted of denaturation
for 15s at 98 $^{\circ}$C, annealing for 20s at 55 $^{\circ}$C, and extension
for 25s at 72$^{\circ}$C. A final extension for five minutes at 72 $^{\circ}$C
was followed by a 16 $^{\circ}$C hold.

\paragraph{} Twelve samples for which the initial PCR reaction produced limited
amplification were subjected to a second PCR reaction. Instead of genomic DNA,
the initial PCR products were used at a reaction volume of 2.0 $\mu$L
(See lab notebook pp. 70-76). For two *ycf1* samples (PVITU 9196 and
PVITU 9837), the PCR yielded two DNA fragments revealed by double banding
following gel electrophoresis. A Thermo Scientific Ultra Agarose Spin Kit was
used to isolate the target DNA following the standard PCR reaction and gel
electrophoresis protocols. The band containing DNA molecules of size equal
to those where PCR resulted in a single band was excised from the gel.

\paragraph{} To isolate DNA from the agarose gel, the excised bands were
measured for mass then placed into spin columns. A volume of gel binding
solution (NaClO4) equal to three times the sample mass was pipetted directly
onto the spin column filters. The samples were incubated for two minutes at 65
$^{\circ}$C, after which the samples were mixed by inverting the spin columns.
An additional incubation of two minutes and sample mixing ensured complete
melting of the gel. The spin columns were centrifuged for 10s at 10,000 rpm
to separate the gel product from DNA. Spin columns were vortexed for 5s to
mix the flow-through. The solution in the collection tube was reloaded to the
spin filters and centrifuged again for 10s at 10,000 rpm. Flow-through was then
discarded and 300 $\mu$L of gel wash solution (Tris, Ethanol) was added to each
spin column. Sample tubes were centrifuged for 10s at 10,000 rpm before
discarding the flow-through and centrifuging for an additional 30s at 10,000
rpm. Spin columns were transferred to clean 1.5 mL collection tubes and 50
$\mu$L ddH2O was added to each spin column. Sample tubes were centrifuged for
30s at 10,000 rpm to elute DNA.

\paragraph{} After DNA isolation from agarose, the following protocol was used
to concentrate the DNA prior to sequencing. 2 $\mu$L 5 M NaCl was added to 49
$\mu$L DNA of each sample. 122.5 $\mu$L ice cold 100% EtOH was then added
before mixing the samples and freezing for 1 hour. The samples were centrifuged
for 10 minutes at 14,000 rpm. Each sample tube was decanted gently to prevent
disrupting the pellet. Samples were then washed with 1 mL of 70% EtOH and
gently mixed. Sample tubes were spun for five minutes at 14,000 rpm before
decanting again. Any remaining ethanol was let to evaporate before
re-suspending in 18 mL ddH2O.

\clearpage

\vspace*{-24pt}

## Gel Electrophoresis

\paragraph{} Success of PCR amplification was verified using a 1.1% agarose
mini gel. A solution of 285 mL 20x SB buffer (boric acid and 10mM sodium
hydroxide, pH 8.5) and 15 $\mu$L ethidium bromide (10 mg/mL) was used as a gel
running buffer. For the gel, 35 mL 20x SB buffer was added to 0.315g agarose.
This mixture was brought to a boil three times, mixing between boils, before
pouring the agar onto a gel tray and refrigerating. Wells were loaded with a
12 $\mu$L mixture of 2.4 $\mu$L SB loading dye (20mM EDTA, 6x SB buffer, 13.5%
Ficoll-400, and 0.5% Orange-G), 8.6 $\mu$L ddH2O, and 1 $\mu$L sample DNA.
Gels were run for 30 to 60 minutes at 130 volts, then washed with deionized
H2O for 5 minutes. A UV illuminator connected to a Kodak EDAS 290
Electrophoresis Documentation and Analysis System was used to document digital
images of the gels.

## DNA Sequencing

\paragraph{} Prior to sequencing, the PCR reaction products were treated with
a mixture consisting of Shrimp Alkaline Phosphatase and Exonuclease I
(78 $\mu$L ddH2O; 2.0 $\mu$L Exonuclease I, 10 U/ $\mu$L; 20 $\mu$L Shrimp
Alkaline Phosphatase, 1 U/$\mu$L) to degrade remnant primers and dNTPs.
A volume of the enzyme mixture equivalent to 1/5 volume of the PCR reaction
(3.8 $\mu$L for *ycf1*, 4.8 $\mu$L for rps and ITS) was added into each PCR
reaction tube. The tubes were then incubated on a thermocycler for 15 min at
37$^{\circ}$C followed by a 15 min incubation at 80$^{\circ}$C.
PCR products were sent to the Iowa State University DNA facility
for sequencing. Samples were processed via Sanger sequencing on an Applied
Biosystems 3730xl DNA Analyzer. Amplification primers (5 mM; rITS-F, rITS-R,
*rps*-F, *rps*-R2, *ycf1*-F, *ycf1*-R; Table \@ref(tab:TablePCRPrimers))
were sent at a volume of seven $\mu$L (15 ng/$\mu$L) for each reaction.
Sequence data were returned as chromatograms from which DNA sequences were
extracted using the software program 4Peaks v1.7.2 [@4Peaks].
Forward and reverse reads for each sample were aligned with Clustal X v2.0
[@Clustal] to verify consistency between sequence reads. Polymorphic sites were
labeled using IUPAC ambiguity codes [@Cornish-Bowden1985]. The sequences were
assembled from chromatograms into multi-FASTA format for downstream analysis.

## Phylogenetic Analysis

\paragraph{} Phylogenetic relationships were estimated by bayesian inference
using two programs, MrBayes v3.2.6 [@Ronquist2012; @Huelsenbeck2001] and
BEAST v2.6.3 [@Bouckaert2019] on the high performance computing platform
CIPRES Science Gateway version 3.3 [@CIPRES]. The Bioconductor package
`Biostrings` [@R-Biostrings] was used to write FASTA subset files from raw
multi-FASTA sequences to account for unequal sequence sampling between targeted
loci. Alignments were calculated individually for each locus from both the total
multi-FASTA sequences (rITS: n=`r length(methods_variables$fasta_msa$rITS)`,
*rps*: n=`r length(methods_variables$fasta_msa$rps)`,
*ycf1*: n=`r length(methods_variables$fasta_msa$ycf1)`) and the subset of
specimens with equal sequencing effort (n=51 for rITS, *rps*, and *ycf1*).
Individual gene trees were estimated by MrBayes for the rITS and *rps* loci
using the full set of aligned sequences (n=72 and n=66, respectively).
A concatenated FASTA file for partitioned multi-locus analysis with MrBayes
was also built from the subset alignments using Biostrings. Alignments were
performed separately for each locus with the full multi-FASTA files
and the multi-FASTA files with the subset of 51 specimens.

\paragraph{} Multiple sequence alignments were calculated from multi-FASTA
files by online MAFFT version 7.306B [@Katoh2013] with the G-INS-i
alignment strategy [@Katoh2005]. This algorithm implements a global alignment
[@Needleman1970], in which flanking sequences are truncated to set the alignment
array domain. Iterative refinement occurs by alignment optimization based on
weighted sum-of-pairs scores [@Gotoh1995]. The 1PAM / k = 2 model, where "1PAM"
indicates an assumption of an average one substitution per 100 sites, was chosen
due to the expected low sequence divergence between closely related species.
For gene tree analyses with MrBayes, pairwise distances were calculated by
p-distance with gaps treated as pairwise deletions using MEGA version 10.1.8
[@Kumar2018, @Stecher2020] to remove identical sequences. Evolutionary model
fits were estimated by jModelTest [@Posada2008] using Bayesian information
criterion to determine the best model for each locus. Sequence alignment gaps
were coded using 2matrix [@Salinas2014], a Perl script that implements
“simple indel coding” as described by @Simmons2000.

\paragraph{} Nexus files output by 2Matrix were partitioned to
differentiate sequence data and coded gaps. The sequence partitions were
assigned a model allowing different rates for transitions and transversions
based on the results of the model test (MrBayes equivalents: ITS = HKY;
*rps* and *ycf1* = T92), while the coded gaps were treated as standard binary
data. The parameters for the distribution shape of among-site rate variation
(shape), nucleotide state frequency (statefreq), and proportion of invariant
sites (pinvar) were unlinked to allow separate estimation. All data were
analyzed for 20,000,000 generations with a sample frequency of 10,000
generations. Two parallel analyses were run, each with 4 chains and a
temperature parameter of 0.04. A third of the trees were discarded as the
burn-in fraction (0.33), and the stop value was set to 0.001.

\paragraph{} Additional NEXUS (*.nex*) files were written from the subset
sequence alignments with the Bioconductor package `ape` [@R-ape]. BEAST analysis
*.xml* files were configured using BEAUti v2.6.3 [@Drummond2012].
A multi-locus partitioned analysis with linked trees included rITS, *rps*, and
*ycf1* sequences from all ingroup taxa (n=49). *Physaria* sensu lato specimens
treated as outgroup taxa in the MrBayes inferences were excluded from BEAST
analyses as inconsistent sampling intensity relative to the ingroup may violate
assumptions of standard time-tree priors [@Drummond2015 p. 98]. Recombinant
sites were checked by 3Seq (build 170612, [@3SEQ]) with none identified.
Identical intra-locus sequences were retained for all taxa and ambiguous sites
were allowed for the nuclear rITS locus. Site models were unlinked and set as
TN93 [@Tamura1993] with empirical frequencies for rITS and JC69 [@Jukes1969]
for both plastid loci *rps* and *ycf1*; substitution rates were estimated for
each locus. A strict clock model was applied to rITS and a linked plastid loci
with a uniform prior. Given the suspected recent divergence of ingroup taxa,
a coalescent constant population tree prior was set with a 1/X population
size parameter. Chain length was set to 50M, logging every 5K trace.
Runs were submitted to CIPRES running BEAST v2.6.3 [@Drummond2007; @Suchard2009].
Output log files were inspected using Tracer v1.7.1 [@Tracer] to verify
parameter convergence by an effective sample size (ESS) over 200.
Three independent runs of 50M states were combined using LogCombiner v2.6.3
[@Bouckaert2019] with 10,000 re-sampled states.
A maximum clade credibility (mcc) tree was calculated from the run output trees
using TreeAnnotator v2.6.3 (*ibid.*) with 20% of the trees discarded as
burn-in and a posterior probability limit of 0.5.

\paragraph{} The geographic history of related species provides a
potentially informative source of information for testing and interpreting
evolutionary hypotheses [@KnowlesMaddison2002].
Phylogeography aims to understand processes responsible for inter- and
intra-specific genetic variation while accounting for potential gene tree
discordance due to differing rates of coalescence (Reviewed by @Knowles2009).
Model-based approaches facilitate statistically rigorous inferences
for addressing phylogeographic questions [@Beaumont2010].
An additional BEAST v1.10.4 analysis was completed using the rITS and rps
alignments (n=60 taxa) with a partition containing bivariate trait data from
decimal degree coordinates. Ancestral geographic locations are inferred by
modeling spatial diffusion using a Cauchy relaxed random walk (RRW)
implemented as a modified Brownian diffusion process with branch-specific
dispersal rates [@Lemey2010; @Pybus2012]. Nucleotide substitution models were
unlinked, using TN93 for nuclear rITS and JC69 for plastid *rps* intron.
Clock models were also unlinked, applying a strict clock model to each
partition, and a constant population size coalescent prior was applied.
The MCMC chain length was set to 100,000,000 and sampled every 10,000
generations. A maximum credibility clade (MCC) tree was calculated from 8,000
trees, discarding 20% as burn-in using TreeAnnotator v1.10.4.

\paragraph{} To account for potential gene tree incongruence arising from
recent lineage divergence, an additional method that explicitly models
incomplete lineage sorting was used to estimate a species tree
[@Heled2010; @Knowles2007]. Included alignments were limited to the rITS
and *rps* intron (n=60 samples each) given the absence of potentially
informative sites (defined as synapomorphic characters) in *ycf1*.
Site and clock model parameters were retained from the partitioned
multi-locus analysis and a yule model was set for the species tree prior.
The multispecies coalescent model was set as a linear population function with
autosomal nuclear ploidy for rITS and non-recombining ploidy for plastid *rps*.
A \*BEAST template was used to assign taxon sets based on the following
successive species hypotheses:
  (1) *a priori* subspecies and a novel species from the Medicine Range,
  (2) lumped *P. saximontanta* subspecies *saximontana* and *dentata*,
  (3) combined *P. saximontana* subsp. as *P. didymocarpa* subsp. *didymocarpa*,
  (4) combined *P. d.* subsp. *lyrata* with *P. d.* subsp. *didymocarpa*
MCMC chains were run for 100M generations, sampling every 10,000.
Maximum clade credibility trees were summarized using TreeAnnotator
from the output species trees discarding 10% of trees as burn-in with a
posterior probability limit of 0.5.

## Source Code

\paragraph{} To plot phylogenetic results, MrBayes consensus and BEAST
maximum clade credibility tree data were read into R by the `treeio` package
[@R-treeio2020]. Inferred phylogenies were plotted with the `ggplot2`
extension `ggtree` [@R-ggtree; @R-ggtree2020], where annotations for duplicate
sequences in the MrBayes analyses were included using `dplyr` and `purrr`.
Multiple sequence alignment were subset to potentially informative sites and
visualized using `ggmsa` [@R-ggmsa].
For project reproducibility, source code and data were maintained
as an R package [@R-packages] utilizing the `usethis` [@R-usethis]
and `devtools` [@R-devtools] packages. A project library with specific package
versioning was maintained using the R package `renv` [@R-renv].
A PDF document and figures from built from source with `bookdown`
[@R-bookdown; @R-bookdown2016], `rmarkdown` [@R-rmarkdown; @R-rmarkdown2018],
`knitr` [@R-knitr; @R-knitr2015], and `tinytex` [@R-tinytex], a $LaTeX$
distribution for typesetting.  Style guide requirements were set with additional
$LaTeX$ packages `titletoc` [@titletoc] for table of contents, `titlesec`
[@titlesec] for section titles, and `titleps` [@titleps] for page style.
The GNU `make` utility v3.81 [@GNUmake] was used
to automate figure generation and document compilation.
References were maintained using Zotero v5.0.93 [@Zotero] with the
Better BibTeX v5.2.86 [@BetterBibTeX] add-on. GitHub repositories were
maintained for the R package source code
(https://github.com/jasonratcliff/thesis)
and `bookdown` manuscript (https://github.com/jasonratcliff/thesis).

\vspace*{20pt}

```{r TableSpeciesTreeHypotheses}

methods_variables$star_beast <-
  system.file("extdata/BEAST", package = "thesis") %>%
  list.files(pattern = "spp_hypotheses.csv", full.names = TRUE) %>%
  readr::read_csv(file = ., col_types = 'ccccc') %>%
  stats::setNames(c("Sample", paste("Hypothesis", 1:4))) %>%
  dplyr::arrange(`Hypothesis 1`, Sample) %>%
  dplyr::mutate(
    dplyr::across(
      .fns = ~ gsub("medicinae", "'medicinae'", x = .x)
    )
  )

if (knit_variables$pandoc == "latex") {
  methods_variables$star_beast <-
    methods_variables$star_beast %>%
    dplyr::mutate(
      Sample = gsub("_", "\\\\_", .data$Sample),
      dplyr::across(
        .cols = dplyr::matches("^Hypothesis"),
        .fns = ~ purrr::map_chr(.x = .x, function(taxa) {
          genus <- stringr::str_extract(
            string = taxa,
            pattern = "^[A-Z]\\."
            )
          species <- stringr::str_extract(
            string = taxa,
            pattern = "(?<=^[A-Z]\\. )'?[a-z]+'?"
            )
          subspecies <- stringr::str_extract(
            string = taxa,
            pattern = "(?<=subsp\\. )[a-z]+$"
            )
          taxon <- paste(
            thesis::italicize(genus, knit_variables$pandoc),
            ifelse(
              test = grepl(pattern = "'[a-z]+'", x = species),
              yes = paste0("‘",
                stringr::str_extract(
                  string = species,
                  pattern = "(?<=')[a-z]+(?=')"),
                "’" 
              ),
              no = thesis::italicize(species, knit_variables$pandoc)
            ),
            ifelse(
              test = !is.na(subspecies),
              yes = paste("\nsubsp.",
                thesis::italicize(subspecies,knit_variables$pandoc)
                ),
              no = ""
            )
          )
          kableExtra::linebreak(x = taxon)
        })
      )
    )
  }

```

```{r TableSpeciesTreeHypothesesLatex}

if (knit_variables$pandoc == "latex") {
  methods_variables$star_beast %>%
    dplyr::slice(1:32) %>%
    knitr::kable(
      x = .,
      format = knit_variables$kable_format,
      escape = FALSE,
      caption = "*BEAST species tree hypotheses.",
      caption.short = "*BEAST species tree hypotheses"
    ) %>%
    kableExtra::kable_styling(
      full_width = FALSE,
      font_size = 8,
      latex_options= "hold_position"
    ) %>%
    kableExtra::row_spec(row = 0, bold = TRUE, font_size = 10) %>%
    kableExtra::column_spec(1, border_left = TRUE) %>%
    kableExtra::column_spec(5, border_right = TRUE) %>%
    kableExtra::collapse_rows(columns = 2:5)
}

```

\clearpage
\captionsetup[table]{labelformat=empty}
\captionsetup[table]{list=no}

```{r TableSpeciesTreeHypothesesLatexContinued}

if (knit_variables$pandoc == "latex") {
  methods_variables$star_beast %>%
    dplyr::slice(33:60) %>%
    knitr::kable(
      x = .,
      format = knit_variables$kable_format,
      escape = FALSE,
      caption = "Table 4. Continued."
    ) %>%
    kableExtra::kable_styling(
      full_width = FALSE,
      font_size = 8,
      latex_options= "hold_position"
    ) %>%
    kableExtra::row_spec(row = 0, bold = TRUE, font_size = 10) %>%
    kableExtra::column_spec(1, border_left = TRUE) %>%
    kableExtra::column_spec(5, border_right = TRUE) %>%
    kableExtra::collapse_rows(columns = 2:5)
}

```

\captionsetup[table]{labelformat=default}
\captionsetup[table]{list=yes}
\setcounter{table}{4}

::: {.content-visible when-format="docx"}

{{< include /appendix/_references.qmd >}}

:::
