\chapter*{REFERENCES}
\addcontentsline{toc}{chapter}{REFERENCES}
\vspace*{-12pt}

\setlength{\parindent}{-4em}
\setlength{\leftskip}{4em}
\setlength{\parskip}{12pt}
\singlespacing

\hspace*{-4em}

Abellán P, Ribera I. 2017. Using phylogenies to trace the geographical signal of diversification. Journal of Biogeography. 44(10):2236-2246.

Aho AV, Kernighan BW, Weinberger PJ. 1988. The AWK programming language. 1st ed. Murray Hill (NJ): AT\&T Bell Laboratories.

Alexander PJ, Rajanikanth G, Bacon CD, Bailey CD. 2007. Recovery of plant DNA using a reciprocating saw and silica-based columns. Mol Ecol Notes. 7(1):5-9.

Bash manual page. [Computer Program]. 2018. München (Germany): Michael Kerrisk; [Cited 2018 August 29].  Available from http://man7.org/linux/man-pages/man1/getopts.1p.html

Becker RA, Wilks AR. 1993. Maps in S. [Internet]. Murray Hill (NJ): AT\&T Bell Laboratories Statistics Research Report; [cited 2018 Apr 27]. Available from http://ect.bell-labs.com/sl/doc/93.2.ps

Becker RA, Wilks AR. 1995. Constructing a geographical database. [Internet]. Murray Hill (NJ): AT\&T Bell Laboratories Statistics Research Report; [cited 2018 Apr 27]. Available from http://ect.bell-labs.com/sl/doc/95.2.ps

Beitz E. 2000. TEXshade: Shading and labeling of multiple sequence alignments using \LaTeX 2$\backepsilon$. Bioinformatics. 16(2):135-139.

Bloomquist, Lemey, Suchard. 2010. Three roads diverged? Routes to phylogeographic inference. Trends in Ecology and Evolution. 25(11):626-632.

Bodenhofer U, Bonatesta E, Horej\v{s}-Kainrath C, Hochreiter S. 2015. msa: An R package for multiple sequence alignment. Bioinformatics. 31(24):3997-3999.

Comai L. 2005. The advantages and disadvantages of being polyploid. Nature Reviews Genetics. 6(11):836-846.

Cornish-Bowden A. 1985. Nomenclature for incompletely specified bases in nucleic acid sequences: recommendations 1984.  Nucleic Acids Research. 13(9):3021-3030.

Darwin. 1859. On the origin of species.

Dayrat B. 2005. Towards integrative taxonomy. Biological Journal of the Linnean Society. 85(3):407-415.

de Queiroz K. 1998. The general lineage concept of species, species criteria, and the process of speciation: A conceptual unification and terminological recommendations. In: Howard DJ, Berlocher SH, editors. Endless forms: species and speciation. New York (NY): Oxford Univ. Press. p. 57–75.

Drabkova L, Kirschner J, Vlcek C. 2002. Comparison of seven DNA extraction and amplification protocols in historical herbarium specimens of Juncaceae. Plant Mol Biol Reporter. 20(2):161-175.


Fairbairns R, Rahtz S, Barroca L. 2016. rotating - rotation tools, including rotated full-page floats [Internet]. Heidelberg (Germany): Comprehensive TeX Archive Network; [cited 2018 April 23]. Available from https://ctan.org/pkg/rotating?lang=en.

Fois M, Fenu G, Cañadas EM, Bacchetta G. 2017. Disentangling the influence of environmental and anthropogenic factors on the distribution of endemic vascular plants in Sardinia. PLoS ONE. 12(8):e0182539.

Gillespie RM, Roderick GK. 2014. Geology and climate drive diversification. Nature. 509(7500):297-298.

GNU bash manual. [Internet]. 2016. Boston (MA): Free Software Foundation; [Cited 2018 August 25]. Available from https://www.gnu.org/software/bash/manual/

Google Inc. 2019. Google Earth [Internet]. Mountain View (CA): Google Inc.; [cited 2019 Sep 29]. Available from https://www.google.com/earth/

Gotoh O. 1995. A weighting system and algorithm for aligning many phylogenetically related sequences. Compu Appl Biosci. 11(5):543-51.

Graham CH, Carnaval AC, Cadena CD, Zamudio KR, Roberts TE, Parra JL, McCain CM, Bowie RCK, Moritz C, Baines SB, Schneider CJ, VanDerWal J, Rahbek C, Kozak KH, Sanders NJ. 2014. The origin and maintenance of montane diversity: Integrating evolutionary and ecological processes. Ecography. 37(8):711-719.

Griekspoor A, Groothuis T. 2015. 4Peaks version 1.7.2. Nucleobytes.

Hey J, Waples RS, Arnold ML, Butlin RK, Harrison RG. 2003. Understanding and confronting species uncertainty in biology and conservation. Trends in Ecology and Evolution. 18(11):597-603.

Hijmans RJ. 2020. raster: Geographic data analysis and ,odeling. R package version 3.0-12. https://CRAN.R-project.org/package=raster

Huelsenbeck JP, Ronquist F. 2001. MRBAYES: Bayesian inference of phylogeny. Bioinformatics. 17(8):754-755.

Isaac NJB, Mallet J, Mace GM. 2004. Taxonomic inflation: Its influence on macroecology and conservation. Trends in Ecology and Evolution. 19(9):464-469.

Kates HR, Johnson MG, Gardner EM, Zerega NJC, Wickett NJ. 2018. Allele phasing has minimal impact on phylogenetic reconstruction from targeted nuclear gene sequences in a case study of \textit{Artocarpus}. American Journal of Botany. 105(3):404-416.

Katoh K, Kuma K, Toh H, Miyata T. 2005. MAFFT version 5: Improvement in accuracy of multiple sequence alignment. Nucleic Acids Research. 33(2):511-518.

Katoh K, Standley DM. 2013. MAFFT multiple sequence alignment software version 7: Improvements in performance and usability.  Mol Biol and Evol. 30(4):772-780.

Kluge J, Worm S, Lange S, Long D, Böhner J, Yangzom R, Miehe G. 2017. Elevational seed plants richness patterns in Bhutan, Eastern Himalaya. J. Biogeography. 44(8):1711-1722.

Knowles LL, Massatti R. 2017. Distributional shifts - not geographic isolation – as a probable driver of montane species divergence. Ecography. 40(12):1475-1485.

Körner, C. 2000. Why are there global gradients in species richness? Mountains might hold the answer. Trends Ecol Evol. 15(12):513–514.

Kumar S, Stecher G, Tamura K. 2016. MEGA7: Molecular evolutionary genetics analysis version 7.0 for bigger datasets. Mol Biol Evol. 33(7):1870-1874.

Laiolo P, Pato J, Obeso JR. 2018. Ecological and evolutionary drivers of the elevational gradient of diversity. Ecology Letters. 21(7):1022-1032.

Larkin MA, Blackshields G, Brown NP, Chenna R, McGettigan PA, McWilliam H, Valentin F, Wallace IM, Wilm A, Lopez R, Thompson JD, Gibson TJ, Higgins DG. 2007. Clustal W and Clustal X version 2.0. Bioinformatics. 23:2947-2948.

Lomolino MV. 2001. Elevation gradients of species-density: Historical and prospective views. Global Ecology and Biogeography. 10(1):3-13.

Maddison WP. 1997. Gene trees in species trees. Syst Biol. 46(3):523-536.

Maddison WP, Knowles LL. 2006. Inferring phylogeny despite incomplete lineage sorting. Syst Biol. 55(1):21-30.

Manton I. 1932. Introduction to the general cytology of the Cruciferae. Annals of Botany. 46(532):509-556.

McCain CM, Grytnes J-A. 2010. Elevational gradients in species richness. In: Encyclopedia of Life Sciences (ELS). Chichester (England): John Wiley and Sons, Ltd.  DOI: 10.1002/9780470015902.a0022548

Miller JS, Venable DL. 2000. Polyploidy and the evolution of gender dimorphism in plants. Science. 289(5488):2335-2338.

Miller MA, Pfeiffer W, Schwartz T. 2010. Creating the CIPRES science gateway for inference of large phylogenetic trees. Paper presented at: GCE. Proceedings of the Gateway Computing Environments Workshop; New Orleans, LA.

Mulligan GA. 1967. Diploid and autotetraploid \textit{Physaria vitulifera} (Cruciferae). Canadian Journal of Botany. 45:183-188.

Myers N, Mittermeier RA, Mittermeier CG, da Fonseca GAB, Kent J. 2000. Biodiversity hotspots for conservation priorities. Nature. 403(6772):853-858.

Needleman S, Wunsch C. 1970. A general method applicable to the search for similarities in the amino acid sequence of two proteins. J Mol Biol. 48(3):443-53.

Noroozi J, Talebi A, Doostmohammadi M, Rumpf SB, Linder HP, Schneeweiss GM. 2018. Hotspots within a global biodiversity hotspot - areas of endemism are associated with high mountain ranges. Scientific Reports. 8(1):1-10.

Padial JM, Castroviejo-Fisher S, Köhler J, Vilá C, Chaparro JC, De la Riva I. 2009. Deciphering the products of evolution at the species level: The need for an integrative taxonomy. Zoologica Scripta. 38(4):431-447.

Padial JM, Miralles A, De la Riva I, Vences M. 2010. The integrative future of taxonomy. Frontiers in Zoology. 7(16):1-14.

Pagès H, Aboyoun P, Gentleman R, DebRoy S. 2019. Biostrings: Efficient manipulation of biological strings. R package version 2.50.2. https://www.bioconductor.org/packages/release/bioc/html/Biostrings.html

Paradis E, Schliep K. 2018. ape 5.0: an environment for modern phylogenetics
and evolutionary analyses in R. Bioinformatics. 35(3):526-528.

Raposo MA, Stopiglia R, Brito GRR, Bockmann FA, Kirwan GM, Gayon J, Dubois A. 2017. What really hampers taxonomy and convservation? A riposte to Garnett and Christidis (2017). Zootaxa. 4317(1):179-184.

Rethwisch, K. 2014. A preliminary systematic analysis of the species of genus \textit{Physaria} (Brassicaceae) of Wyoming. Honors Program Theses. Paper 105.
http://scholarworks.uni.edu/hpt/105

Rokas A, Carroll SB. 2005. More genes or more taxa? The relative contribution of gene number and taxon number to phylogenetic accuracy. Mol Biol Evol. 22(5):1337-1344.

Rollins RC, Banerjee UC. 1979b. Trichome patterns in \textit{Physaria} (Cruciferae). Publ. Bussey Inst. Harvard Univ. 1979:65-77.

Ronquist F, Teslenko M, Van der Mark P, Ayres DL, Darling A, Höhna S, Larget B, Liu L, Suchard MA, Huelsenbeck JP. 2012. MrBayes 3.2: Efficient Bayesian inference and model choice across a large model space. Syst Biol. 61(3):539-542.

Rostagi S, Liberles DA. 2005. Subfunctionalization of duplicated genes as a transistion state to neofunctionalization. BMC Evol Biol. 5(28):1-7.

Salinas NR, Little DP. 2014. 2Matrix: A utility for indel coding and phylogenetic matrix concatenation. Applications in Plant Sci. 2(1): 1300083.

Simmons MP, Ochoterena H. 2000. Gaps as characters in sequence-based phylogenetic analyses. Syst Biol. 49(2):369-381.

Simpson GG. 1951. The species concept. Evolution. 5(4):285–298.

Shaffer BH, Thomson RC. 2007. Delimiting species in recent radiations. Syst Biol. 56(6):896-906.

Steinbauer MJ, Field R, Grytnes, J-A, Trigas P, Ah-Peng C, Attorre F, Birks HJB, Borges PAV, Cardoso P, Chou C-H, De Sanctis M, de Sequeira MM,  Duarte MC, Elias RB, Fern\v{a}ndez-Palacios JM, Gabriel R, Gereau RE, Gillespie RG, Greimler J, Harter DEV, Huang T-J, Irl SDH, Jeanmonod D, Jentsch A, Jump AS, Kueffer C, Nogu\v{e} S, Otto R, Price J, Romeiras MM, Strasberg D, Stuessy T, Svenning J-C, Vetaas OR, Beierkuhnlein C. 2016. Topography-driven isolation, speciation and a global increase of endemism with elevation. Global Ecol Biogeogr. 25(9): 1097-1107.

Steinbauer MJ, Irl SDH, Beierkuhnlein C. 2013. Elevation-driven ecological isolation promotes diversification on Mediterranean islands. Acta Oecologica. 47(1):52-56.


Tamura K, Peterson D, Peterson N, Stecher G, Nei M, Kumar S. 2011. MEGA5: Molecular evolutionary genetics analysis using maximum likelihood, evolutionary distance, and maximum parsimony methods. Mol Biol and Evol. 28(10): 2731-2739.

Tamura K, Stecher G, Peterson D, Filipski A, Kumar S. 2013. MEGA6: Molecular evolutionary genetics analysis version 6.0. Mol Biol and Evol. 30(12): 2725-2729.

Templeton AR. 1989. The meaning of species and speciation: A genetic perspective. In Otte D, Endler JA, editors. Speciation and its consequences. Sunderland (MA): Sinauer, p. 3-27.

Thiele K, Yeates D. 2002. Tension arises from duality at the heart of taxonomy. Nature. 419(6905):337.

Thomson SA, Pyle RL, Ahyong ST, Alonso-Zarazaga M, Ammirati J, Araya JF, et al. 2018. Taxonomy based on science is necessary for global conservation. PLoS Biol. 16(3):e2005075.

Trigas P, Panitsa M, Tsiftsis S. 2013. Elevational gradient of vascular plant species richness and endemism in Crete - the effect of post-isolation mountain uplift on a continental island system. PLoS ONE 8(3): e59425. doi:10.1371/journal.pone.0059425

Vetaas OR, Grytnes J-A. 2002. Distribution of vascular plant species richness and endemic richness along the Himalayan elevation gradient in Nepal. Global Ecology and Biogeography. 11(4):291-301.

Wang LG, Lam TTY, Xu S, Dai Z, Zhou L, Feng T, Guo P, Dunn CW, Jones BR, Bradley T, Zhu H, Guan Y, Jiang Y, Yu G. treeio: an R package for phylogenetic tree input and output with richly annotated and associated data. 2019. Molecular Biology and Evolution. 37(2):599-603.

Weber WA, Brewbaker JL. 1950. \textit{Physaria vitulifera}, a tetraploid species of Cruciferae. Series in Biology. 3:24-28.

Weigelt P, Kreft H. 2013. Quantifying island isolation – insights from global patterns of insular plant species richness. Ecography. 36(4):417-429.


Wiens JJ. 2004. What is speciation and how should we study it. The American Naturalist. 163(6):914-923.

Wünschiers R. 2013. Computational biology: A practical introduction to biodata processing and analysis with Linux, MySQL, and R. 2nd ed. Berlin (Heidelberg): Springer-Verlag.

Xie Y. 2018. bookdown: Authoring books and technical documents with R markdown. R package version 0.9. https://cran.r-project.org/package=bookdown

Yu G, Smith D, Zhu H, Guan Y, Lam TT-Y. 2017. ggtree: An R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution. 8(1):28-36.

Zachos FE. 2016. An annotated list of species concepts. In: Species concepts in biology. Switzlerand: Springer International Publishing. p. 77-96.

Zachos FE. 2018. (New) Species concepts, species delimitation and the inherent limitations of taxonomy. Journal of Genetics. 97(4):811-815.

\setlength{\parindent}{0em}
\setlength{\leftskip}{0em}
\setlength{\parskip}{6pt}
\doublespacing
