# METHODS {#METHODS}

\vspace*{-26pt}

```{r methodsSectionSpecimens}
knitr_section(knitr_title = "Specimen Sampling and Tissue Sources",
              knitr_type = "section")
```

\paragraph{} Specimens of *Physaria* sensu stricto were requested from the following herbaria: RM, NY, GH, MONT, MONTU, IDS, F, UC, RSA-POM, MO, UTC, and US.  Species distribution maps were established using geographic data collected from herbarium specimen labels and coordinate data converted into decimal degrees.  Montana State University’s Graphical Locator for TRS data (Wefald 2003) was used to approximate collection locations to the center of a section from specimen Township, Range, and Section data.  When coordinate data were not provided, Google Earth (Google Inc. 2015) was used to estimate coordinates to half-minute precision based on the specimen location description.  Specimens with ambiguous locations (e.g. a river or canyon) were omitted from mapping.  Coordinates were plotted using R to generate projections of species distributions (Appendix - Source Code).

\paragraph{} There has been considerable debate in the literature regarding the extent to which density of taxon sampling and gene number effect the accuracy of phylogenetic inference.  Simulation studies by Maddison and Knowles (2006) found for recent divergences, sampling more individuals increased the accuracy of species tree inference.  In such cases, increased taxon sampling may improve the number of detectable gene coalescence events between sister taxa. In contrast, simulations by Rokas and Carroll (2005) found increasing the number of genetic loci analyzed improved the accuracy of analysis while taxon sampling had either no statistically significant effect or a slight negative correlation.  Therefore, samples were chosen from across each species’ range to broadly represent populations within their distribution and multiple loci were chosen for each individual (Table `r knitr_table2()`).

\clearpage

```{r methodsTable2DnaSpecimens}
# LaTeX specimen table
if (knitr_chunk == "latex") {
  read.csv("data/tables/methods-DNA-specimens.csv",
           header = TRUE, row.names = NULL, as.is = TRUE,
           check.names = FALSE, stringsAsFactors = FALSE,
           colClasses = "character") %>%
    dplyr::mutate(`Voucher Information` = sapply(
      `Voucher Information`, function(voucher) {
        gsub("[A-Z]\\. ?", "", voucher) %>% gsub("&", "and", x = .)
        })) %>%
    kable(format = knitr_chunk, escape = FALSE,
          caption = "DNA Extraction Herbarium Specimens.") %>%
    kable_styling(full_width = FALSE, font_size = 10,
                  latex_options= "hold_position") %>%
    row_spec(row = 0, bold = TRUE, font_size = 12) %>%
    column_spec(1, border_left = TRUE, width = "3cm") %>%
    column_spec(3, border_right = TRUE) %>%
    collapse_rows(columns = 1) %>%
    footnote(general =
               paste0("Requested herbarium specimens used for extraction of ",
                      "genomic DNA.  Voucher information includes record ",
                      "collector and collection number while locality reports ",
                      "the county and state for the collection.  Specimens ",
                      "with a single * represent vouchers with DNA sequence ",
                      "data obtained by K. Rethwisch (2014) while a double ** ",
                      "indicate specimens with sequence data obtained by Dr. ",
                      "O'Kane."), general_title = "",
             threeparttable = TRUE, escape = FALSE)
}
```

\clearpage

<!--- Clear caption formatting to prevent automatic labelling -->
\captionsetup[table]{labelformat=empty}
<!--- Prevent automatic indexing of table entry in List of Tables -->
\captionsetup[table]{list=no}

```{r methodsTable2DnaSpecimensContd, tab.cap = NULL}

# For supressing the latex caption above, reference here:
# https://stackoverflow.com/questions/31182147/suppress-automatic-table-name-and-number-in-an-rmd-file-using-xtable-or-knitr?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa

# For suppressing the list of tables entry, reference here:
# https://texblog.org/2010/05/02/excluding-items-from-list-of-figures-list-of-tables/

if (knitr_chunk == "latex") {
  read.csv("data/tables/methods-DNA-specimens-2.csv",
           header = TRUE, row.names = NULL, as.is = TRUE,
           check.names = FALSE, stringsAsFactors = FALSE,
           colClasses = "character") %>%
    dplyr::mutate(`Voucher Information` = sapply(
      `Voucher Information`, function(voucher) {
        gsub("[A-Z]\\. ?", "", voucher) %>% gsub("&", "and", x = .)
        })) %>%
    kable(format = knitr_chunk, escape = FALSE,
          caption = "Table 2. Continued.") %>%
    kable_styling(full_width = FALSE, font_size = 10,
                  latex_options= "hold_position") %>%
    row_spec(row = 0, bold = TRUE, font_size = 12) %>%
    column_spec(1, border_left = TRUE, width = "3cm") %>%
    column_spec(3, border_right = TRUE) %>%
    collapse_rows(columns = 1)
}
```

<!--- Reset caption options and table counter -->
\captionsetup[table]{labelformat=default}
\captionsetup[table]{list=yes}
\setcounter{table}{2}

```{r methodsTable2DnaSpecimensHtml}
# HTML specimen table
if (knitr_chunk == "html") {
  bind_rows(
    read.csv("data/tables/methods-DNA-specimens.csv",
           header = TRUE, row.names = NULL, as.is = TRUE, check.names = FALSE,
           stringsAsFactors = FALSE, colClasses = "character"),
    read.csv("data/tables/methods-DNA-specimens-2.csv",
             header = TRUE, row.names = NULL, as.is = TRUE,
             check.names = FALSE, stringsAsFactors = FALSE,
             colClasses = "character")
    ) %>%
    dplyr::mutate(taxa = sapply(Taxon, function(taxon) {
      gsub("\\\\textit\\{", "\\*", taxon) %>%
        gsub("\\}$", "\\*", x = .) %>%
        gsub("\\} ssp.", "\\*  ssp.", x = .)
  })) %>%
    select(c(taxa, `Voucher Information`, Locality)) %>%
    dplyr::rename(Taxon = taxa) %>%
    kable(longtable = TRUE, format = knitr_chunk, escape = FALSE,
          caption = "DNA Extraction Herbarium Specimens") %>%
    kable_styling(full_width = TRUE, font_size = 10) %>%
    row_spec(row = 0, bold = TRUE, font_size = 12) %>%
    column_spec(1, border_left = TRUE, bold = TRUE) %>%
    column_spec(2, width = "8cm") %>%
    column_spec(3, width = "4cm", border_right = TRUE) %>%
    collapse_rows(columns = 1) %>%
    footnote(general =
               paste0("Requested herbarium specimens used for extraction of ",
                      "genomic DNA.  Voucher information includes record ",
                      "collector and collection number while locality reports ",
                      "the county and state for the collection.  Specimens ",
                      "with a single * represent vouchers with DNA sequence ",
                      "data obtained by K. Rethwisch (2014) while a double ** ",
                      "indicate specimens with sequence data obtained by Dr. ",
                      "O'Kane."), general_title = "",
             threeparttable = TRUE, escape = FALSE)
}
```

\vspace*{12pt}

\paragraph{} Taxon sampling was determined by species distribution patterns, with broader sampling from more widespread species.  For the narrowly distributed Wyoming state endemics *P. dornii* and *P. condensata*, three individuals were sampled from each taxa.  For Wyoming state endemics *P. eburniflora* and *P. saximontana* ssp. *saximontana*, five individuals from each taxa were included.  Four individuals of *P. saximontana* ssp. *dentata* were sampled broadly from Montana.  Due to the limited number of available specimens, only three specimens identified as *P. didymocarpa* ssp. *lanata* in Wyoming and three identified as *P. didymocarpa* ssp. *lyrata* in Idaho were included.  These two *P. didymocarpa* subspecies have narrow distributions and are thought to be state endemics of Wyoming and Idaho, respectively.  Five individuals were sampled from *P. didymocarpa* ssp. *didymocarpa* in Wyoming and Montana.

\paragraph{} For the regional endemics *P. brassicoides* and *P. integrifolia*, six and five individuals were sampled, respectively.  For *P. vitulifera*, three individuals from Colorado and four putative specimens with annotations as *P. vitulifera* from Carbon County, Wyoming were included.  The Carbon County population is of interest due to the questionable status of this species’ distribution in the state of Wyoming (Heidel & O’Kane 2012).  Finally, twelve individuals of the widespread *P. acutifolia* from Utah, Colorado, and Wyoming were included.

\paragraph{} Genomic DNA was extracted from one of two sources: basal leaf tissue taken from loaned herbarium specimens (ca. 20 mg) or silica bead-dried leaf tissue from specimens collected in June and August of 2014 from Wyoming, Idaho, and Montana with Dr. S. L. O’Kane Jr. and Mr. M. Horsch.  Dried leaf tissues were stored at -20 $^{\circ}$C prior to extraction.  DNA was extracted from only a single individual specimen to minimize noise in the sequence data.  Sequence data from a preliminary systematic analysis of Physaria sensu stricto by Kayla Rethwisch (2014) in addition to data previously obtained by Dr. O'Kane were also included in this study.

```{r methodsSectionExtractions}
knitr_section(knitr_title = "DNA Extraction",
              knitr_type = "section")
```

\paragraph{} Genomic DNA was purified using a Qiagen DNeasy Plant Mini Kit.  Prior to the extraction protocol, leaf tissue was prepared using one of three methods.  Firstly, approximately 20 mg dried plant tissue was combined with ca. 250 mg washed sea sand and ground to a fine powder using a mortar and pestle.  Secondly, a modified approach of Alexander et al. (2007) was used where ca. 20 mg of plant tissue was combined in a 2.0 mL microcentrifuge tube with three 4.5 mm zinc plate BBs and two 2.3 mm stainless steel beads.  Samples were attached to a reciprocating saw and the saw was operated for 90s at medium until leave tissue was pulverized.  Finally, a Retsch MM400 Bead Mill was used in place of the reciprocating saw with the same bead, BB and tissue combination.  Plant tissue was disrupted for 45s at 20 hertz resulting in a fine ground powder.

\paragraph{} Once the plant tissue was prepared, Buffer AP1 (400 $\mu$L) and RNase A (4 $\mu$L) were added.  Samples were vortexed and incubated for 30 minutes at 65 $^{\circ}$C.  Sample microcentrifuge tubes were inverted after five, 15, and 25 minutes during incubation.  Following incubation, Buffer P3 (130 $\mu$L) was added.  Samples were then hand mixed and incubated on ice for five minutes.  The resultant lysate was centrifuged for five minutes at 14,000 rpm.  Lysate supernatant was transferred to a QIAshredder spin column and centrifuged for two minutes at 14,000 rpm.  Flow-through was then transferred to a new 1.5 mL microcentrifuge tube and mixed by pipette with 1.5x volume Buffer AW1.  650 $\mu$L of the resulting mixture was transferred to a DNeasy mini spin column and centrifuged for one minute at 9,000 rpm.  Flow-through was discarded before adding the remaining mixture to the column, and centrifuging again for one minute at 9,000 rpm.

\paragraph{} Spin columns were then placed in new 2.0 mL collection tubes.  Buffer AW2 (500 $\mu$L) was added and the tubes were centrifuged for one minute at 10,000 rpm.  Flow-through was discarded before adding additional Buffer AW2 (500 $\mu$L) to each column.  Spin columns were centrifuged for two minutes at 14,000 rpm before transfer to new 1.5 mL centrifuge tubes. Preheated Buffer AE (75 $\mu$L at 65 $^{\circ}$C) was added to elute DNA from the columns.  Samples were incubated for 10 minutes at room temperature before centrifuging for one minute at 9,000 rpm.  Additional preheated Buffer AE (75 $\mu$L at 65 $^{\circ}$C) was added to each column.  Samples were incubated at room temperature for ten minutes before centrifuging an additional minute at 9,000 rpm.  Modifications to the standard Qiagen DNA extraction protocol followed recommendations made by Drabkova et al. (2002).  Time for the initial incubation period was extended from 10 to 30 minutes to ensure complete cellular lysis.  Additionally, the volume of AE buffer used to elute DNA from the spin column was reduced to 75 $\mu$L from 100 $\mu$L.  The incubation time for column elution was also extended from five minutes to 10 minutes.

```{r methodsSectionPcrIts}
knitr_section(knitr_title = "rITS PCR",
              knitr_type = "section")
```

\paragraph{} All Polymerase Chain Reaction (PCR) was completed using a Biometra® TProfessional Basic Thermocycler Gradient with individual or 8-tube strip 0.25 mL PCR microcentrifuge tubes.  Individual reactions for the rITS region had a total volume of 25 $\mu$L consisting of 13.85 $\mu$L distilled H2O (ddH2O), 2.5 $\mu$L 10X Reaction Buffer (500 mM KCl; 100 mM Tris-HCl, pH 8.8; 17.5 mM MgCl2; 0.5% v/v Triton-X; 5% v/v DMSO), 2.5 $\mu$L rITS-F primer (2 $\mu$M, Table \@ref(tab:methodsTable3DnaPrimers)), 2.5 $\mu$L rITS-R primer (2 $\mu$M, Table \@ref(tab:methodsTable3DnaPrimers)), 2.5 $\mu$L 8 mM dNTP mix (10 $\mu$L 100 mM dATP; 10 $\mu$L 100 mM dGTP; 10 $\mu$L 100 mM dCTP; 10 $\mu$L 100 mM dTTP; 5 $\mu$L 1 M Tris-HCl, pH 7.9; 455 $\mu$L ddH2O), 0.15 $\mu$L Taq DNA Polymerase (5 U/$\mu$L) and 1.0 $\mu$L sample DNA.  Lid temperature for thermocycling was set to 99 $^{\circ}$C.  An initial DNA denaturation period of two minutes at 94 $^{\circ}$C was followed by 30 cycles of denaturation, primer annealing, and DNA extension.  Each cycle consisted of denaturation for 45s at 94 $^{\circ}$C, primer annealing for one minute at 60 $^{\circ}$C, and extension for one minute at 72 $^{\circ}$C.  After 30 cycles, a final extension of five minutes at 72 $^{\circ}$C was followed by a 16 $^{\circ}$C hold.

\vspace*{24pt}

```{r methodsTable3DnaPrimers}

# Store character vector for type of knitr chunk (i.e. 'latex' | 'html')
# Add Caption formatted by knitr chunk type.
knitr_chunk <- opts_knit$get("rmarkdown.pandoc.to")

# Read in specimen data .csv file from Phys_DNA/ subdirectory.
dna_prime <- read.csv("Phys_DNA/Seq-Tables/methods-DNA-primers.csv", 
                      header = TRUE, row.names = NULL, as.is = TRUE,
                      check.names = FALSE, stringsAsFactors = FALSE, 
                      colClasses = "character")

kable(dna_prime, format = knitr_chunk, align = "l", escape = FALSE,
      caption = "Primer Sequences for PCR Reactions.") %>%
  kable_styling(full_width = FALSE, font_size = 10,
                latex_options= "hold_position") %>%
  row_spec(row = 0, bold = TRUE, font_size = 12) %>%
  column_spec(1, border_left = TRUE) %>%
  column_spec(3, border_right = TRUE) %>%
  footnote(general = "DNA primers for PCR reactions.", 
           footnote_as_chunk = TRUE, general_title = "", 
           escape = TRUE, threeparttable = TRUE)

```

\vspace*{12pt}

```{r methodsSectionPcrRps}
knitr_section(knitr_title = "*rps* Intron PCR",
              knitr_type = "section")
```

\paragraph{} For the *rps* intron, individual reactions had a total volume of 25 $\mu$L consisting of 11.25 $\mu$L ddH2O, 5.0 $\mu$L 5x Phusion HF Buffer (proprietary, Thermo Fisher Scientific), 2.5 $\mu$L *rps*-F (2 $\mu$M, Table (Table \@ref(tab:methodsTable3DnaPrimers))), 2.5 $\mu$L *rps*-R2 (2 $\mu$M, Table (Table \@ref(tab:methodsTable3DnaPrimers))), 2.5 $\mu$L dNTPs, 0.25 $\mu$L Phusion Hot Start II High-Fidelity DNA Polymerase (Thermo Fisher Scientific), and 1.0 $\mu$L sample DNA.  Lid temperature for thermocycling was set to 99 $^{\circ}$C.  An initial 1 minute denaturation period at 98 $^{\circ}$C was followed by 30 cycles.  Each cycle consisted of denaturation for 10s at 98 $^{\circ}$C, annealing for 25s at 66.8 $^{\circ}$C, and extension for 30s at 72 $^{\circ}$C.  A final extension for seven minutes at 72 $^{\circ}$C was followed by a 16 $^{\circ}$C hold.

```{r methodsSectionPcrYcf1}
knitr_section(knitr_title = "*ycf1* PCR",
              knitr_type = "section")
```

\paragraph{} The total volume for *ycf1* reactions was 20 $\mu$L.  Due to inconsistencies in the success of PCR amplification, the volume of sample DNA in the PCR reactions ranged from 1.0 $\mu$L to 2.0 $\mu$L.  An equivalent decrease in the volume of ddH2O, from 8.8 $\mu$L to 7.8 $\mu$L was used to maintain a total 20 $\mu$L volume.  In addition, the reaction mixture included 4.0 $\mu$L Phusion HF 5x buffer, 2.0 $\mu$L dNTPs, 2.0 $\mu$L *ycf1*-F (2 $\mu$M, Table (Table \@ref(tab:methodsTable3DnaPrimers))), 2.0 $\mu$L *ycf1*-R (2 $\mu$M, Table (Table \@ref(tab:methodsTable3DnaPrimers))), and 0.2 $\mu$L Phusion DNA polymerase.  Lid temperature for thermocycling was set to 99 $^{\circ}$C.  An initial denaturation period for one minute at 98 $^{\circ}$C was followed by 35 cycles.  Each cycle consisted of denaturation for 15s at 98 $^{\circ}$C, annealing for 20s at 55 $^{\circ}$C, and extension for 25s at 72$^{\circ}$C.  A final extension for five minutes at 72 $^{\circ}$C was followed by a 16 $^{\circ}$C hold.

\paragraph{} Twelve samples for which the initial PCR reaction produced limited amplification were subjected to a second PCR reaction.  Instead of genomic DNA, the initial PCR products were used at a reaction volume of 2.0 $\mu$L (See lab notebook pp. 70-76).  For two *ycf1* samples (PVITU 9196 and PVITU 9837), the PCR yielded two DNA fragments revealed by double banding following gel electrophoresis.  A Thermo Scientific Ultra Agarose Spin Kit was used to isolate the target DNA following the standard PCR reaction and gel electrophoresis protocols.  The band containing DNA molecules of size equal to those where PCR resulted in a single band was excised from the gel.

\paragraph{} To isolate DNA from the agarose gel, the excised bands were measured for mass then placed into spin columns.  A volume of gel binding solution (NaClO4) equal to three times the sample mass was pipetted directly onto the spin column filters.  The samples were incubated for two minutes at 65 $^{\circ}$C, after which the samples were mixed by inverting the spin columns.  An additional incubation of two minutes and sample mixing ensured complete melting of the gel.  The spin columns were centrifuged for 10s at 10,000 rpm to separate the gel product from DNA.  Spin columns were vortexed for 5s to mix the flow-through.  The solution in the collection tube was reloaded to the spin filters and centrifuged again for 10s at 10,000 rpm.  Flow-through was then discarded and 300 $\mu$L of gel wash solution (Tris, Ethanol) was added to each spin column.  Sample tubes were centrifuged for 10s at 10,000 rpm before discarding the flow-through and centrifuging for an additional 30s at 10,000 rpm.  Spin columns were transferred to clean 1.5 mL collection tubes and 50 $\mu$L ddH2O was added to each spin column.  Sample tubes were centrifuged for 30s at 10,000 rpm to elute DNA.

\paragraph{} After DNA isolation from agarose, the following protocol was used to concentrate the DNA prior to sequencing.  2 $\mu$L 5 M NaCl was added to 49 $\mu$L DNA of each sample.  122.5 $\mu$L ice cold 100% EtOH was then added before mixing the samples and freezing for 1 hour.  The samples were centrifuged for 10 minutes at 14,000 rpm.  Each sample tube was decanted gently to prevent disrupting the pellet.  Samples were then washed with 1 mL of 70% EtOH and gently mixed.  Sample tubes were spun for five minutes at 14,000 rpm before decanting again.  Any remaining ethanol was let to evaporate before re-suspending in 18 mL ddH2O.

```{r methodsSectionPcrGel}
knitr_section(knitr_title = "Gel Electrophoresis",
              knitr_type = "section")
```

\paragraph{} Success of PCR amplification was verified using a 1.1% agarose mini gel.  A solution of 285 mL 20x SB buffer (boric acid and 10mM sodium hydroxide, pH 8.5) and 15 $\mu$L ethidium bromide (10 mg/mL) was used as a gel running buffer. For the gel, 35 mL 20x SB buffer was added to 0.315g agarose.  This mixture was brought to a boil three times, mixing between boils, before pouring the agar onto a gel tray and refrigerating.  Wells were loaded with a 12 $\mu$L mixture of 2.4 $\mu$L SB loading dye (20mM EDTA, 6x SB buffer, 13.5% Ficoll-400, and 0.5% Orange-G), 8.6 $\mu$L ddH2O, and 1 $\mu$L sample DNA.  Gels were run for 30 to 60 minutes at 130 volts, then washed with de-ionized H2O for 5 minutes.  A UV illuminator connected to a Kodak EDAS 290 Electrophoresis Documentation and Analysis System was used to document digital images of the gels.

```{r methodsSectionSeq}
knitr_section(knitr_title = "DNA Sequencing",
              knitr_type = "section")
```

\paragraph{} Prior to sequencing, the PCR reaction products were treated with a mixture consisting of Shrimp Alkaline Phosphatase and Exonuclease I (78 $\mu$L ddH2O; 2.0 $\mu$L Exonuclease I, 10 U/ $\mu$L; 20 $\mu$L Shrimp Alkaline Phosphatase, 1 U/$\mu$L) to degrade remnant primers and dNTPs.  A volume of the enzyme mixture equivalent to 1/5 volume of the PCR reaction (3.8 $\mu$L for *ycf1*, 4.8 $\mu$L for rps and ITS) was added into each PCR reaction tube.  The tubes were then incubated on a thermocycler for 15 min at 37$^{\circ}$C followed by a 15 min incubation at 80$^{\circ}$C.

\paragraph{} PCR products were sent to Iowa State’s DNA facility for sequencing.  Samples were processed via Sanger sequencing on an Applied Biosystems 3730xl DNA Analyzer.  Amplification primers (5 mM; rITS-F, rITS-R, *rps*-F, *rps*-R2, *ycf1*-F, *ycf1*-R) were sent at a volume of seven $\mu$L (15 ng/$\mu$L) for each reaction.  Sequence data were returned as chromatograms from which DNA sequences were extracted using the software program 4Peaks (Greikspoor and Groothuis 2015).  Forward and reverse reads for each sample were aligned with Clustal X (Larkin et al. 2007) to verify consistency between sequence reads.  Polymorphic sites were labeled using IUPAC ambiguity codes (Cornish-Bowden 1985).  The sequences were assembled from chromatograms into multi-FASTA format for downstream analysis.

```{r methodsSectionPhylo}
knitr_section(knitr_title = "Phylogenetic Analysis",
              knitr_type = "section")
```

\paragraph{} Multiple sequence alignments were calculated from multi-FASTA files by online MAFFT version 7.306B (Katoh and Standley 2013) with the G-INS-i alignment strategy (Katoh et al. 2005).  This algorithm implements a global alignment (Needleman and Wunsch 1970), in which flanking sequences are truncated to set the alignment array domain.  Iterative refinement occurs by optimizing the alignment based on weighted sum-of-pairs scores (Gotoh 1995). The 1PAM / k = 2 model, where "1PAM" indicates an assumption of an average one substitution per 100 sites, was chosen due to the expected low sequence divergence between closely related *Physaria* species.  After aligning loci separately, sequences were exported into a wrapped multi-FASTA format from the MAFFT server.  These files were then processed with a series of Unix shell scripts (Wünschiers 2013) before concatenating loci for multi-locus Bayesian inference (Appendix C).  Extra sequences were removed from the multi-fasta files such that only specimens with data for all three loci were included for multiple sequence alignments of individual loci for concatenated analysis.  Local analyses were run with Mac OS X version 10.11.6, Unix OS Darwin version 13.4.0, and Terminal version 2.6.1 on an x86_64 processor.

\paragraph{} MEGA 7.0.18 (Kumar et al. 2016; Tamura et al. 2013; Tamura et al. 2011) was used to test individual loci alignments for the evolutionary model with the highest likelihood using an automatic neighbor-joining tree with gaps treated as partial deletions with a 95% site coverage cutoff.  To detect identical sequences, MEGA was used to measure the number of differences between sequences.  Sequences with distances of 0.0 were verified as identical using a custom R function (Appendix - Source Code), as the pairwise number of distance calculation exludes indices with heterozygous ambiguity codes and alignment gaps.  Sequence alignment gaps were coded using 2matrix (Salinas and Little 2014), a Perl script that implements “simple indel coding” as described by Simmons and Ochoterena (2000).

\paragraph{} Bayesian analyses were executed using Mr. Bayes version 3.2.6 (Ronquist et al. 2012; Huelsenbeck and Ronquist 2001) on the CIPRES Science Gateway version 3.3 (Miller et al. 2010).  Data were partitioned to differentiate sequence data and coded gaps.  The sequence partitions were assigned a model allowing different rates for transitions and transversions based on the results of the model test (ITS = HKY; *rps* and *ycf1* = T92), while the coded gaps were treating as standard binary data.  The parameters for the distribution shape of among-site rate variation (shape), nucleotide state frequency (statefreq), and proportion of invariant sites (pinvar) were unlinked to allow separate estimation.  All data were analyzed for 20,000,000 generations with a sample frequency of 1,000 generations.  Two parallel analyses were run, each with 4 chains and a temp parameter of 0.04.  A third of the trees were discarded as the burn-in fraction (0.33), and the stop value was set to 0.002.
