# RESULTS {#RESULTS}

\vspace*{-26pt}

```{r resultsSectionBayesInf}
knitr_section(knitr_title = "Phylogenetic Analysis",
              knitr_type = "section")
```

\paragraph{} Phylogenetic analyses with MrBayes were completed for two individual loci, rITS and *rps* intron.  Pairwise distances of the *ycf1* loci revealed a high degree of sequence similarity with 36 of 51 specimens having identical genotypes.  Individual gene tree analyses also indicated a marked degree of sequence homogeneity, with seven groupings of identical genotypes for rITS and nine such groupings for *rps* intron.  In addition to the individual gene trees, a concatenated analysis was performed using samples with sequence data from rITS, *rps* intron, and *ycf1*, for which two groupings of identical combined genotypes were observed.

<!-- <INSERT BEAST WRITEUP> -->

```{r resultsSubsectionItsSeq}
knitr_section(knitr_title = "ITS Sequence Analysis",
              knitr_type = "subsection")
```

\paragraph{} Sequences for the rITS loci ranged from 581 to 583 base pairs with an alignment length of 586 base pairs.  There were seven instances of identical genotypes, often between interspecific specimens according to previous annotations (Table \@ref(tab:resultsTable4IdenticalSeqITS)).  These included one group of 20 specimens (**genotype 1**) with four specimens each of previously annotated *P. dornii* and *P. condensata* from southwestern Wyoming; two specimens of *P. integrifolia* from Teton County and one from Lincoln County; seven specimens of *P. acutifolia* from southern Wyoming and western Colorado; and three specimens from the Big Horn Mountains (*Nelson 74315*, *O'Dea 32*, *Ratcliff and O'Kane 48*).  Three specimens of *P. vitulifera* from central Colorado and one of *P. acutifolia* from central Wyoming also had identical rITS sequences (**genotype 2**).  Another grouping of four specimens included three of putative *P. didymocarpa* subsp. *lanata* from Johnson County, Wyoming and one of *P. brassicoides* from Powder River County, Montana (**genotype 3**).  Additional groupings include two specimens from central Wyoming, *Smith 551* and *Ratcliff and O'Kane 50* (**genotype 4**); two specimens from Idaho previously annotated as *P. didymocarpa* subsp. *lyrata* (**genotype 5**); and two sets of two *P. eburniflora* specimens (**genotypes 6** and **7**).

\clearpage

```{r resultsTable4IdenticalSeqITS, fig.width=8}

tbl_phylo_rITS <-
  phylo_tbl(bayes_file = paste0("data/8.bayes/rITS-infile.nex.con.tre"),
            specimen_records = total_physaria)

phylo_multi_tbl(phylo_tbl_obj = tbl_phylo_rITS) %>%
  phylo_kable(tbl_multi_node = .,
              kable_caption =
                    paste0("$Physaria$ DNA specimens with identical ",
                           "genotypes for the rITS loci."))

```

\clearpage

\paragraph{} As previously described, no species of were recovered as monophyletic groups in the rITS gene tree.  *P. obcordata*, *P. fendlerii* and *P. argyraea* were strongly supported as outgroup taxa (P=1.000).  Single gene Bayesian inference of the rITS loci showed low resolution overall, with a strongly supported 19 node soft polytomy (P=1.000, Figure \@ref(fig:resultsFigure3rITSPhyloTree)).  Inlcuded in the polytomy were specimens previously annotated as *P. vitulifera*, *P. saximontana* subsp. *dentata*, *P. eburnifora*, *P. didymocarpa* subsp. *didymocarpa*, *P. d.* subsp. *lanata*, *P. d.* subsp. *lyrata*, *P. dornii*, *P. condensata*, *P. brassicoides* and *P. acutifolia*.  In the polytomy, **Genotype 1** consisted of all sampled individuals previously annotated as *Physaria condensata* (*Ratcliff and O'Kane, Jr. 40*; *Ratcliff and O'Kane, Jr. 41*; *O'Kane, Jr. 3787*; *Fertig 16765*) and *P. dornii* (*Ratcliff and O'Kane, Jr. 42*; *Refsdal and Refsdal 4376*; *Fertig 17503*; *Fertig 17506*) in addition to three *P. integrifolia* specimens (*Nelson 19288*, *Nelson and Scott 68918*, and *Fertig 17493*).  Six specimens of *P. acutifolia* from southern Wyoming and western Colorado in addition to three specimens from near Big Horn Canyon also had the identical genotype.

\paragraph{} A moderately supported polytomy (P=0.741) included four specimens of *P. didymocarpa* subsp. *didymocarpa*, two specimens of *P. saximontana* subsp. *dentata*, two specimens of *P. s.* subsp. *saximontana*, and single specimens of *P. didymocarpa* subsp. *lanata* and *P. d.* subsp. *lyrata*, *P. brassicoides*, a *P. acutifolia* collection from Utah (*Holmgren and Holmgren 14222*), and putative *P. vitulifera* from Carbon County (*Ratcliff and O'Kane, Jr. 34*).  Within the second polytomy, two nested clades of low and moderate support included *P. brassicoides* (P=0.958) and *P. saximontana* subsp. *dentata* (P=0.507).  Two additional clades recovered with high support included *P. saximontana* subsp. *saximontana* and  *P. eburniflora* (P=0.995), and *P. integrifolia* (P=0.982).

\clearpage

```{r resultsFigure3rITSPhyloTree, fig.cap=knitr_cap, fig.width=8, fig.asp=1.3}

knitr_cap <- "Consensus tree from Baysian inference of the rITS gene tree."

phylo_ggplot(phylo_tbl_obj = tbl_phylo_rITS,
             phylo_layout = "rectangular", plot_title = "rITS Phylogeny",
             legend_col = 1, x_expand = 0.025, legend_text = 10) %>%
  lemon::reposition_legend(aplot = ., position = "top left")

```

\clearpage
\vspace*{-26pt}

```{r chapterSubsectionRpsSeq}
knitr_section(knitr_title = "*rps* Intron Sequence Analysis",
              knitr_type = "subsection")
```

\paragraph{} Sequences for the *rps* intron ranged from 802 to 852 base pairs with an aligned length of 889.  A high degree of sequence homogeneity was observed in the *rps* intron alignment (Table \@ref(tab:resultsTable5IdenticalSeqRps)).  Fourteen identical sequences (**genotype 1**) included specimens previously identified as *P. didymocarpa* subsp. *didymocarpa* (n=2), *P. d.* subsp. *lanata* (n=2), *P. saximontana* susp. *saximontana* (n=2), *P. s.* subsp. *dentata* (n=2), *P. brassicoides* (n=3), and *P. acutifolia* (n=3).  These samples were primarily from populations of the Big Horn and Absorokee ranges of Wyoming and Montana, with outliers in the Big Belt mountains of Montana, northeastern Wyoming, and south-central Wyoming.  Another grouping of identical sequences included specimens primarily from western Montana and east-central Idaho (**genotype 2**).  These specimens were previously annotated as *P. didymocarpa* subsp. *didymocarpa* (n=2), *P. d.* subsp. *lyrata* (n=2), *P. saximontana* subsp. *dentata* (n=2), and *P. acutifolia* (n=1).

\vspace*{24pt}

```{r resultsTable5IdenticalSeqRps, fig.width=8}

tbl_phylo_rps <-
  phylo_tbl(bayes_file = "data/8.bayes/rps-infile.nex.con.tre",
            specimen_records = total_physaria)
tbl_multi_rps <- phylo_multi_tbl(phylo_tbl_obj = tbl_phylo_rps)

if (knitr_chunk == "html") {
 phylo_kable(tbl_multi_node = tbl_multi_rps,
            kable_caption =
              paste0("$Physaria$ DNA specimens with identical ",
                     "genotypes for the $rps$ intron loci."))
} else if (knitr_chunk == "latex") {
  phylo_kable(tbl_multi_node = tbl_multi_rps[1:21, ],
            kable_caption =
              paste0("$Physaria$ DNA specimens with identical ",
                     "genotypes for the $rps$ intron loci."))
}

```

\clearpage

<!--- Clear caption formatting to prevent automatic labelling -->
\captionsetup[table]{labelformat=empty}
<!--- Prevent automatic indexing of table entry in List of Tables -->
\captionsetup[table]{list=no}

```{r resultsTable5IdenticalSeqRpsCont, fig.width=8}

if (knitr_chunk == "latex") {
  phylo_kable(tbl_multi_node = tbl_multi_rps[22:47, ],
            kable_caption = "Table 5. Continued.")
}

if (knitr_chunk == "latex") {
  cat("\\captionsetup[table]{labelformat=default}")
  cat("\\captionsetup[table]{list=yes}")
  cat("\\setcounter{table}{5}")
}

```

\vspace*{12pt}

\paragraph{} A third genotype consisted of three specimens with recent annotations as *P. saximontana*, *P. eburniflora*, and *P. acutifolia* from Hot Springs county near the Wind River Range, and two specimens of *P. acutifolia* from western Sublette County, Wyoming and Rio Blanco, Colorado.  This group (**genotype 3**) also included *Dorn 10105*, a specimen from the Medicine Bow Range with a previous annotation as *P. vitulifera*.  A fourth group (**genotype 4**) included two other specimens (*Ratcliff and O'Kane 34*, *Fertig and Welp 19075*) from the Medicine Bow Range.  The fourth group also included collections of *P. integrifolia* from Lincoln County, Wyoming; *P. acutifola* from Grand County, Utah; and *P. brassicoides* from Converse County, Wyoming.  Five specimens with identical sequences from Lincoln County, Wyoming (**genotype 5**) included two collections of *P. dornii*, two of *P. condensata*, and one collection of *P. integrifolia*.  Three specimens of *P. didymocarpa* subsp. *lanata* from the Big Horn Mountains and one of *P. acutifolia* from Uinta County, Wyoming comprised another group (**genotype 6**).  One specimen of *P. saximontana* subsp. *dentata* from Glacier County, Montana and one of *P. eburniflora* from Carbon County, Wyoming had identical sequences (**genotype 7**).  Lastly, two *P. eburniflora* collections from Fremont County, Wyoming (**genotype 8**) and two *P. vitulifera* collections from Colorado (**genotype 9**) comprised two additional genotypes.

\paragraph{} Individual gene tree analysis for the *rps* intron did not recover monophyletic groups of species as previously described (Figure \@ref(fig:resultsFigure4rpsPhyloTree)).  Here, a 13 node polytomy was observed (P=1.000).  Unresolved taxa included *P. vitulifera* of Colorado, *P. eburniflora*, *P. acutifolia*, *P. saximontana* subsp. *saximontana*, *P. s.* subsp. *dentata*, *P. didymocarpa* subsp. *didymocarpa*, *P. d.* subsp. *lyrata*, *P. d.* subsp. *lanata*, *P. integrifolia*, one specimen from the Medicine Bow Range, and *P. obcordata*.  Both *P. fendlerii* and *P. argyraea* were recovered as outgroups with high support (P=1.000).

\paragraph{} For the *rps* intron, a higher degree of sequence heterogeneity was observed for putative *P. condensata*, *P. dornii*, and *P. integrifolia* relative to the rITS gene tree.  Here, a clade of moderate support (P=0.860) was recovered containing three specimens of *P. dornii*, three of *P. condensata*, two of *P. integrifolia*, and one *P. acutifolia* specimen from Montezuma County, Colorado.  One specimen of *P. condensata* was recovered in a separate, highly supported clade (P=0.980) also containing *P. saximontana* subsp. *saxmimontana* (n=3), *P. s* subsp. *dentata* (n=2), *P. didymocarpa* subsp. *didymocarpa* (n=2), *P. d.* subsp. *lanata* (n=2), *P. brassicoides* (n=1), and *P. acutifolia* (n=1).  Two other clades with moderate to high support were also recovered, one with Colorado *P. vitulifera* specimens (n=2, P=0.826) and another with *P. acutifolia* (n=1), *P. brassicoides* (n=1), *P. integrifolia* (n=1), and two specimen from the Medicine Bow Range (*Fertig and Welp 19075*, *Ratcliff and O'Kane, Jr. 34*; P=0.996).

```{r resultsFigure4rpsPhyloTree, fig.cap=knitr_cap, fig.width=8, fig.asp=1.3}

knitr_cap <- "Consensus tree from Baysian inference of *rps* intron gene tree."

ggtree_rps <-
  phylo_ggplot(phylo_tbl_obj = tbl_phylo_rps, phylo_layout = "rectangular",
               plot_title = expression(paste(italic("rps"),
                                             " intron Phylogeny")),
               legend_col = 1, x_expand = 0.015, legend_text = 10) +
  ylim(y = c(0, 35))

lemon::reposition_legend(aplot = ggtree_rps, position = "top left")

```

\clearpage

\vspace*{-26pt}

```{r chapterSubsectionYcf1Seq}
knitr_section(knitr_title = "*ycf1* Sequence Analysis",
              knitr_type = "subsection")
```

\paragraph{} Sequencing of the *ycf1* locus indicated a low mutation rate for that marker in this group of taxa.  Specimens sampled from each of the putative species considered in this study had identical sequences, with 36 individuals having an identical genotype (Table \@ref(tab:resultsTable6IdenticalSeqYcf1)).  Those showing some degree of heterogeneity for this loci included specimes of *P. acutifolia* (n=3), *P. brassicoides* (n=2), *P. integrifolia* (n=2), *P. eburniflora* (n=2), *P. condensata* (n=1), *P. didymocarpa* subsp. *didymocarpa* (n=1), *P. d.* subsp. *lanata* (n=1), and one specimen from the Medicine Bow Range (*Fertig 16713*).  Due to the high degree of homogeneity observed for this marker, an individual gene tree analysis was not carried out.

\vspace*{24pt}

```{r resultsTable6IdenticalSeqYcf1}

knitr_cap <-
  paste0("$Physaria$ DNA specimens with identical genotypes",
         " for the $ycf1$ loci.")

# Format ycf1 identical sequence .csv files into split tables.
ycf1_id_seqs <- read.csv("data/6.MEGA/ycf1_identical_seqs.csv",
                         header = TRUE, row.names = NULL, as.is = TRUE,
                         check.names = FALSE, stringsAsFactors = FALSE,
                         colClasses = "character") %>%
  dplyr::mutate(., Collector =
                    map_chr(.[["Collector"]], function(collector) {
                      gsub("[A-Z]\\. ?", "", collector) %>%
                        gsub("&|with", "and", x = .)})) %>%
  dplyr::select(., Genotype, Species, State, Collector, `Collection Number`)

if (knitr_chunk == "html") {
  phylo_kable(tbl_multi_node = ycf1_id_seqs,
              kable_caption = knitr_cap)
} else if (knitr_chunk == "latex") {
  phylo_kable(tbl_multi_node = ycf1_id_seqs[1:23, ],
              kable_caption = knitr_cap)
}

```

\clearpage

<!--- Clear caption formatting to prevent automatic labelling -->
\captionsetup[table]{labelformat=empty}
<!--- Prevent automatic indexing of table entry in List of Tables -->
\captionsetup[table]{list=no}

```{r resultsTable6IdenticalSeqYcf1Cont}

if (knitr_chunk == "latex") {
  phylo_kable(tbl_multi_node = ycf1_id_seqs[24:36, ],
              kable_caption = "Table 6. Continued.")
}

if (knitr_chunk == "latex") {
  cat("\\captionsetup[table]{labelformat=default}")

  cat("\\captionsetup[table]{list=yes}")

  cat("\\setcounter{table}{6}")
}

```

\vspace*{6pt}

```{r chapterSubsectionMultiLocus}
knitr_section(knitr_title = "Concatenated Multi-locus Analysis",
              knitr_type = "subsection")
```

\paragraph{} Concatenation of the three sequenced loci yielded an alignment with greater overall heterogeneity relative to any individual gene alignment.  Here, only two identical genotypes were observed, one from two specimens previously identified as *P. dornii* and *P. condensata*, and another comprised of two specimens previously annotated as *P. didymocarpa* subsp. *lyrata* (Table \@ref(tab:resultsTable7IdenticalSeqML)).  Although higher posterior probabilities were observed for three clades in the concatenated analysis, 14 nodes were still unresolved as a polytomy (P=0.777, Figure \@ref(fig:resultsFigure5MultiLocusPhyloTree)).  One node with high support (P=0.966) recovered three specimens previously annotated as *P. condensata*, two previously as *P. dornii*, in addition to one specimen of *P. integrifolia* (*Ratcliff and O'Kane Jr., 43*) and another of *P. acutifolia* (*Moore 5199*).

\vspace*{24pt}

```{r resultsTable7IdenticalSeqML, fig.width=8}

tbl_phylo_ml <-
  phylo_tbl(bayes_file = "data/8.bayes/ml-infile.nex.con.tre",
            specimen_records = total_physaria)

phylo_multi_tbl(phylo_tbl_obj = tbl_phylo_ml) %>%
  phylo_kable(tbl_multi_node = .,
              kable_caption =
                paste0("$Physaria$ DNA specimens with identical sequences",
                       " after concatenation."))

```

```{r resultsFigure5MultiLocusPhyloTree, fig.cap=knitr_cap, fig.width=8, fig.asp=1.3}

knitr_cap <- paste0("Consensus tree from Baysian inference of",
                    " concatenated multi-locus data.")

ggtree_ml <-
  phylo_ggplot(tbl_phylo_ml, plot_title = "Multi-locus Phylogeny",
               phylo_layout = "rectangular", legend_col = 1, x_expand = 0.025,
               legend_text = 10)

lemon::reposition_legend(aplot = ggtree_ml, position = "top left")

```

\clearpage

\paragraph{} Another clade recovered with high posterior probability (P=0.837) included specimens previously annotated as *P. saximontana* subsp. *saximontana* (n=3), *P. s.* subsp. *dentata* (n=2), *P. didymocarpa* subsp. *lanata* (n=2), *P. brassicoides* (n=3), *P. acutifolia* (n=2), and one specimen of *P. condensata* (Figure \@ref(fig:resultsFigure5MultiLocusPhyloTree)).  A third clade (P=0.966) included two specimens from Carbon County, Wyoming (*Fertig and Welp 19075*, *Fertig 16713*), *P. acutifolia* (n=1) from Utah,  *P. brassicoides* (n=1) from northeastern Wyoming, and *P. integrifolia* (n=1) from western Wyoming.  Two additional clades included one of two *P. vitulifera* specimens (P=0.867) and another clade (P=0.999) of *P. saximontana* subsp. *saximontana* (n=1) and *P. eburniflora* specimen (n=1).  In the multi-locus polytomy, unresolved specimens were previously identified as *P. acutifolia* (n=4), *P. didymocarpa* subsp. *didymocarpa* (n=3), *P. d.* subsp. *lyrata* (n=1), *P. d.* subsp. *lanata* (n=1), and *P. eburniflora* (n=2), *P. saximontana* subsp. *dentata* (n=1), and a specimen from Carbon County, Wyoming (*Dorn 10105*).  An additional clade with poor support (P=0.504) included *P. didymocarpa* subsp. *lyrata* (n=2) and *P. brassicoides* (n=1).


```{r chapter3_1ResultsCleanup}
# Workspace cleanup
rm(ycf1_id_seqs)
```

