# RESULTS {#RESULTS}

\vspace*{-26pt}
\section{Phylogenetic Analysis}

```{r resultsSectionBayesInf, results = "asis"}

knitr_chunk_html("## Phylogenetic Analysis")

```

\paragraph{} Phylogenetic analyses using Bayesian inference were completed for two individual loci, rITS and *rps* intron. A concatenated analysis  including these two loci in addition to *ycf1* was also carried out.  Individual gene tree analyses revealed a high degree of sequence homogeneity, with seven groupings of identical genotypes for the rITS loci and nine for *rps*.  Additionally, the *ycf1* loci showed the highest degree of overall sequence similarity with 34 of 49 specimens having identical genotypes.

\subsection{ITS Sequence Analysis}

```{r resultsSubsectionItsSeq, results = "asis"}

knitr_chunk_html("### ITS Sequence Analysis")

```

\paragraph{} Sequences for the rITS loci ranged from 581 to 583 base pairs with an alignment length of 586 base pairs.  There were seven instances of identical genotypes, often between interspecific specimens according to previous annotations (Table \@ref(tab:resultsTable1IdenticalSeqITS)).  These included one group of 19 specimens (genotype 1) with four specimens each of previously annotated *P. dornii* and *P. condensata* from southwestern Wyoming; two specimens of *P. integrifolia* from Teton County, Wyoming; six specimens of *P. acutifolia* from southern Wyoming and western Colorado; and three specimens from the Big Horn Mountains (Nelson 74315, O'Dea 32, Ratcliff and O'Kane 48).  Three specimens of *P. vitulifera* from central Colorado and one of *P. acutifolia* from central Wyoming also had identical rITS sequences (genotype 2).  Another grouping of four specimens included three of putative *P. didymocarpa* subsp. *lanata* from Johnson County, Wyoming and one of *P. brassicoides* from Powder River County, Montana (genotype 3).  Additional groupings include two specimens from central Wyoming, *Smith 551* and *Ratcliff and O'Kane 50* (genotype 4); two specimens from Idaho previously annotated as *P. didymocarpa* subsp. *lyrata* (genotype 5); and two sets of two *P. eburniflora* specimens (genotypes 6 and 7).  

\clearpage

```{r resultsTable1IdenticalSeqITS, warning=FALSE}

# Generate ggtree object base for further species specific tree annotation.
bayes_ITS <- "Phys_DNA/Seq-Analysis/7.Bayes/rITS-single-1.1_infile.nex.con.tre"
bayes_ITS_tree <- tree_plot(consensus_tree_file = bayes_ITS)

# Format latex kable for identical DNA sequences.
kable_format(label_resolve_csv = "Phys_DNA/Seq-Analysis/7.Bayes/rITS-single-1.1_label-resolve.csv")

latex_kable(id_seq_csv = "Phys_DNA/Seq-Analysis/7.Bayes/rITS-single-1.1_label-resolve-table.csv",
            kable_caption = "$Physaria$ DNA specimens with identical genotypes for the rITS loci.")

```

\clearpage

\paragraph{} Single gene Bayesian inference of the rITS loci showed low resolution, with a weakly supported (P=0.538) 17 node polytomy (Figure \@ref(fig:resultsFigure1rITSPhyloTree)).  As previously described, no species were recovered as monophyletic groups in the rITS gene tree.  Inlcuded in the polytomy of this tree were specimens previously annotated as *P. vitulifera*, *P. saximontana* subsp. *saximontana*, *P. s.* subsp. *dentata*, *P. eburnifora*, *P. didymocarpa* subsp. *didymocarpa*, *P. d.* subsp. *lanata*, *P. d.* subsp. *lyrata*, *P. dornii*, *P. condensata*, *P. brassicoides* and *P. acutifolia*.  Two specimens of *P. brassicoides* and one of *P. didymocarpa* subsp. *lanata* were recovered as a sister group to the polytomy-containing clade (P=0.969), while *P. obcordata*, *P. fendlerii* and *P. argyraea* were strongly supported as outgroup taxa (P=1.0)

\paragraph{} In the polytomy, **Genotype 1** consisted of all sampled individuals previously annotated as *Physaria condensata* (*Ratcliff and O'Kane, Jr. 40*; *Ratcliff and O'Kane, Jr. 41*; *O'Kane, Jr. 3787*; *Fertig 16765*) and *P. dornii* (*Ratcliff and O'Kane, Jr. 42*; *Refsdal and Refsdal 4376*; *Fertig 17503*; *Fertig 17506*) in addition to two *P. integrifolia* specimens (*Nelson 19288*, *Nelson and Scott 68918*).  Six specimens of *P. acutifolia* from southern Wyoming and western Colorado in addition to three specimens from near Big Horn Canyon also had this identical sequence.  

\paragraph{} A low support clade (P=0.562) including five specimens of *P. didymocarpa* subsp. *didymocarpa*, four specimens of *P. brassicoides*, three specimens of *P. saximontana* subsp. *dentata*, two specimens of *P. s.* subsp. *saximontana*, and one specimen of *P. d.* subsp. *lyrata* was recovered.  This clade also included a *P. acutifolia* collection from Utah (*Holmgren and Holmgren 14222*) and a collection from the Medicine Bow Range (*Ratcliff and O'Kane, Jr. 34*).  Two specimens of *P. saximontana* subsp. *saximontana* and one of *P. eburniflora* were recovered as a clade with low support (P=0.561) while another two *P. eburniflora* specimens (**Genotype 7**) and one *P. integrifolia* were recovered as a separate clade (P=0.729).  

\clearpage

```{r resultsFigure1rITSPhyloTree, fig.width=8, fig.asp=1.3, fig.cap=knitr_cap, warning=FALSE}

knitr_cap <- "Consensus tree from Baysian inference of the rITS gene tree."
tree_grid(bayes_ITS_tree)

```

\clearpage
\vspace*{-26pt}
\subsection{\textit{rps} Intron Sequence Analysis}

```{r chapterSubsectionRpsSeq}

knitr_chunk_html("### *rps* Intron Sequence Analysis")

```

\paragraph{} Sequences for the *rps* intron ranged from 802 to 852 base pairs with an aligned length of 889.  A high degree of sequence homogeneity was observed in the *rps* intron alignment (Table \@ref(tab:resultsTable2IdenticalSeqRps)).  Fourteen identical sequences (genotype 1) included specimens previously identified as *P. didymocarpa* subsp. *didymocarpa* (n=2), *P. d.* subsp. *lanata* (n=2), *P. saximontana* susp. *saximontana* (n=2), *P. s.* subsp. *dentata* (n=2), *P. brassicoides* (n=3), and *P. acutifolia* (n=3).  These samples were primarily from populations of the Big Horn and Absorokee ranges of Wyoming and Montana, with outliers in the Big Belt mountains of Montana, northeastern Wyoming, and south-central Wyoming.  Another grouping of identical sequences included specimens primarily from western Montana and east-central Idaho (Genotype 2).  These included specimens previously annotated as *P. didymocarpa* subsp. *didymocarpa* (n=2), *P. d.* subsp. *lyrata* (n=2), *P. saximontana* subsp. *dentata* (n=2), and one specimen of *P. acutifolia* from Colorado.

\vspace{36pt}

```{r resultsTable2IdenticalSeqRps, warning=FALSE}

# Generate ggtree object base for further species specific tree annotation.
bayes_rps <- "Phys_DNA/Seq-Analysis/7.Bayes/rps-single-1.1_infile.nex.con.tre"
bayes_rps_tree <- tree_plot(consensus_tree_file = bayes_rps,
                            x_adjust_factor = 6, x_expand = 0.02,
                            legend_y_pos = c(0, 1)) + 
   expand_limits(y = 2)

# Format latex kable for identical DNA sequences.
kable_format(label_resolve_csv = "Phys_DNA/Seq-Analysis/7.Bayes/rps-single-1.1_label-resolve.csv",
             table_split = c(2, 9))

latex_kable(id_seq_csv = "Phys_DNA/Seq-Tables/rps-single-1.1_label-resolve-table1.csv",
            kable_caption = "$Physaria$ DNA specimens with identical genotypes for the $rps$ intron loci.")

```

\clearpage

<!--- Clear caption formatting to prevent automatic labelling -->
\captionsetup[table]{labelformat=empty}
<!--- Prevent automatic indexing of table entry in List of Tables --> 
\captionsetup[table]{list=no}

```{r resultsTable2IdenticalSeqRps2}

latex_kable(id_seq_csv = "Phys_DNA/Seq-Tables/rps-single-1.1_label-resolve-table2.csv",
            kable_caption = "Table 5. Continued.")

```

<!--- Reset caption options and table counter -->
\captionsetup[table]{labelformat=default}
\captionsetup[table]{list=yes}
\setcounter{table}{5}

\vspace{36pt}

\paragraph{} A third genotype consisted of three specimens (previously *P. saximontana*, *P. eburniflora*, and *P. acutifolia*) from Hot Springs county near the Wind River Range, and two specimens of *P. acutifolia* from western Sublette County, Wyoming and Rio Blanco, Colorado.  This group (genotype 3) also included *Dorn 10105*, a specimen from the Medicine Bow Range with a previous annotation as *P. vitulifera*.  A fourth group (genotype 4) included two other specimens (*Ratcliff and O'Kane 34*, *Fertig and Welp 19075*) from the Medicine Bow Range.  Additionally, this group included collections of *P. integrifolia* from Lincoln County, Wyoming; *P. acutifola* from Grand County, Utah; and *P. brassicoides* from Converse County, Wyoming.  Five specimens with identical sequences from Lincoln County, Wyoming (Genotype 5) included two collections of *P. dornii*, two of *P. condensata*, and one collection of *P. integrifolia*.  Three specimens of *P. didymocarpa* subsp. *lanata* from the Big Horn Mountains and one of *P. acutifolia* from Uinta County, Wyoming comprised another group (genotype 6).  One specimen of *P. saximontana* subsp. *dentata* from Glacier County, Montana and one of *P. eburniflora* from Carbon County, Wyoming also had identical sequences (Genotype 7).  Lastly, two *P. eburniflora* collections from Fremont County, Wyoming (Genotype 8) and two *P. vitulifera* collections from Colorado (Genotype 9) comprised two additional genotypes.

\paragraph{} Individual gene tree analysis for the *rps* intron did not recover monophyletic groups of species as previously described (Figure \@ref(fig:resultsFigure2rpsPhyloTree)).  Here, a 13 node polytomy was observed.  Unresolved taxa included *P. vitulifera* of Colorado, *P. eburniflora*, *P. acutifolia*, *P. saximontana* subsp. *saximontana*, *P. s.* subsp. *dentata*, *P. didymocarpa* subsp. *didymocarpa*, *P. d.* subsp. *lyrata*, *P. d.* subsp. *lanata*, *P. integrifolia*, *P. obcordata*, and one specimen from the Medicine Bow Range.  Both *P. fendlerii* and *P. argyraea* were recovered as outgroups with high support (P=1.0).

\paragraph{} For the *rps* intron, a higher degree of sequence heterogeneity was observed for putative *P. condensata*, *P. dornii*, and *P. integrifolia* relative to the rITS gene tree.  Here, a clade of moderate support (P=0.848) was recovered containing three specimens of *P. dornii*, three of *P. condensata*, two of *P. integrifolia*, and one *P. acutifolia* specimen from Montezuma County, Colorado.  One specimen of *P. condensata* was recovered in a separate, highly supported clade (P=0.987) also containing *P. saximontana* subsp. *saxmimontana*, *P. s* subsp. *dentata*, *P. didymocarpa* subsp. *didymocarpa*, *P. d.* subsp. *lanata*, *P. brassicoides*, and one specimen of *P. acutifolia*.  Two other clades with moderate to high support were also recovered, one with two Colorado *P. vitulifera* specimens (P=0.736) and another with *P. acutifolia*, *P. brassicoides*, *P. integrifolia*, and one specimen from the Medicine Bow Range (P=0.996).  

\clearpage

```{r resultsFigure2rpsPhyloTree, fig.cap=knitr_cap, fig.width=6.6, fig.asp=1.3, warning=FALSE}

knitr_cap <- "Consensus tree from Baysian inference of *rps* intron gene tree."

tree_grid(bayes_rps_tree)

```

\clearpage

\subsection{ycf1 Sequence Analysis}

```{r chapterSubsectionYcf1Seq, results = "asis", eval=FALSE}

knitr_chunk_html("### ycf1 Sequence Analysis")

```

```{r resultsTable3IdenticalSeqYcf1, eval=FALSE}

# Read in specimen data .csv file from Phys_DNA/ subdirectory.
# In the .csv file, escape character "\n" is quoted.
# Once read, grep sub to replace the surrounding quotes.
seq_ycf1 <- read.csv("Phys_DNA/table-ycf1-seqs.csv", 
                    header = TRUE, row.names = NULL, 
                    check.names = FALSE, stringsAsFactors = FALSE,
                    allowEscapes = TRUE)

# Store character vector for type of knitr chunk (i.e. 'latex' | 'html')
# Add Caption formatted by knitr chunk type.
knitr_chunk <- opts_knit$get("rmarkdown.pandoc.to")

if (knitr_chunk == "latex") { 
  
  # Add Caption formatted by knitr chunk type
  knitr_caption <- "\\textit{Physaria} specimens with identifcal
                    \\textit{rps} intron sequence genotypes."
  
} else if (knitr_chunk == "html") {
  
  knitr_caption <- "*Physaria* specimens with identifcal
                    *rps* intron sequence genotypes."
  
  # Format italics for hmtl knitr table replacing LaTeX function calls.
  seq_ycf1$Species <- gsub("\\\textit\\{", "\\*", seq_ycf1$Species)
  seq_ycf1$Species <- gsub("\\}$", "\\*", seq_ycf1$Species)
  seq_ycf1$Species <- gsub("\\} ssp.", "\\*  ssp.", seq_ycf1$Species)
  seq_ycf1$Collector <- gsub("\\\\&", "&", seq_ycf1$Collector)
  
}

# knitr::kable with additional kableExtra:: options
kable(seq_ycf1, format = knitr_chunk, escape = FALSE, row.names = FALSE, 
      caption = knitr_caption) %>%
  kable_styling(full_width = FALSE, font_size = 10,
                  latex_options= "hold_position") %>%
  row_spec(row = 0, bold = TRUE, font_size = 12) %>%
  column_spec(1, border_left = TRUE) %>%
  column_spec(5, border_right = TRUE, width = "2.35cm") %>%
  collapse_rows(columns = 1:3) %>%
  kableExtra::footnote(general = "$Physaria$ DNA specimens with identical
           genotypes for the $rps$ intron loci.", 
           general_title = "", threeparttable = TRUE, escape = FALSE)


```

\clearpage

```{r resultsFigure3MultiLocusPhyloTree, fig.cap=knitr_cap, fig.width=6.9, fig.asp=1.3, warning=FALSE}

knitr_cap <- "Consensus tree from Baysian inference of concatenated multi-locus data."

# Generate ggtree object base for further species specific tree annotation.
bayes_ml <- "Phys_DNA/Seq-Analysis/7.Bayes/ml-1.1_infile.nex.con.tre"
bayes_ml_tree <- tree_plot(consensus_tree_file = bayes_ml, 
                           x_expand = 0.03)
tree_grid(bayes_ml_tree)

```
