---
title: "data-raw README"
output: rmarkdown::github_document
bibliography: "../docs/thesis/appendix/references.bib"
link-citations: true
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE
)

knitr::opts_knit$set(
  eval.after = "fig.cap"
)

suppressPackageStartupMessages({
  library(thesis)
  library(Biostrings)
  library(DiagrammeR)
  library(fs)
  library(glue)
  library(knitr)
  library(magrittr)
  library(pkgdown)
  library(purrr)
  library(rprojroot)
  library(stringr)
  library(tibble)
})

# Note: phantomjs required for HTML rendering of of Diagrammer::mermaid()
# - To install v2.1.1 via Homebrew:
# $ brew tap homebrew/cask
# $ brew cask install phantomjs

# TODO IDEA .Rmd flag
```

[//]: # (TODO <3/27/23>: Optimize FASTA splitting pipeline | Nextflow)

```{r links}
links <- list()
links$url <-
  pkgdown::as_pkgdown(
    pkg = find_root(criteria$is_pkgdown_project)
  )$repo$url$source

links$extdata <- glue::glue("{links$url}inst/extdata/FASTA")

links$fasta <-
  list(
    subset = "2.subset-fastas",
    single = "3.alignments-single",
    multi = "3.alignments-subset",
    loci = "3.multi-locus",
    script = "fasta-concat.R"
  ) %>%
    purrr::map(
      .x = .,
      .f = function(x)
        glue::glue("{links$url}data-raw/sequencing/{x[[1]]}")
    )

links$data <-
  list(
    mega = "MEGA",
    runs = "Runs"
  ) %>%
    purrr::map(
      .x = .,
      .f = function(x)
        glue::glue("{links$url}data-raw/Bayes/{x[[1]]}")
    )
```

# DNA Sequence Analysis

## Fasta Concatenation

Multi-FASTA files for sequenced loci were assembled from sequence
chromatograms and deposited in the [`inst/extdata/FASTA`](`r links$extdata`)
subdirectory. Here, `FASTA` headers are formatted with two fields `accession`
and `locus` following ">". A single whitespace separates the sample accession
from the name of the locus, for example: `>PACUT_48 rITS` or `>PACUT_12821 rps`.
Each `FASTA` file contains sequences sampled from a single genetic locus,
where the locus is indicated in the second field of the `FASTA` headers.

```{r fastaFiles}
#| echo: true
list.files(
  path = fs::path_package(package = "thesis", "extdata/FASTA")
)
```

The `Biostrings` package [@R-Biostrings] was used to read in `FASTA` files as
`DNAStringSet` objects, an S4 class inheriting the `XString` subclass. Briefly,
a list is assigned with elements containing a single `DNAStringSet` read from
each `FASTA` file. The intersecting set of sequence headers among all `FASTA`
files is identified from the extracted `DNAStringSet` names attributes.
New subsets for each locus were indexed by the set of specimens sequenced
for all loci. Untrimmed, single locus `FASTA` files from
[`inst/extdata/FASTA`](`r links$extdata`)
and `FASTA` files filtered to common specimens in
[`data-raw/sequencing/sequencing/2.subset-fastas`](`r links$fasta$subset`)
were aligned using MAFFT version 7.306B [@Katoh2013].
The G-INS-i alignment [@Katoh2005] with iterative refinement and
1PAM / k=2 nucleotide scoring matrix were set as alignment
parameters. Aligned `FASTA` files were deposited in the
[`data-raw/sequencing/3.alignments-single`](`r links$fasta$single`) and
[`data-raw/sequencing/3.alignments-subset`](`r links$fasta$multi`)
subdirectories. For the subset of samples with DNA sequence data from all three
sample loci, a concatenated multi-FASTA file was compiled using the
[`data-raw/sequencing/fasta-concat.R`](`r links$fasta$script`) R script,
written to [`data-raw/sequencing/3.multi-locus`](`r links$fasta$loci`).

```{r diagFASTA}
DiagrammeR::mermaid('
graph TD
  %% Sequencing Workflow
  S1["inst/extdata/FASTA"]-->S2(fasta-subsets.R)
  S2-->S3["sequencing/2.subset-fastas/"]
  S1-->S4{MAAFT}
  S3-->S4
  S4-->S5["sequencing/3.alignments-single/"]
  S4-->S6["sequencing/3.alignments-subset/"]
  S6-->S7(fasta-concat.R)
  S7-->S8["sequencing/3.multi-locus/"]
')
```

## MEGA / MrBayes

Distance matrices were calculated from the full single-locus and subset
multi-locus alignment FASTA files to identify identical DNA sequences using
MEGA version 10.1.8 [@Kumar2018; @Stecher2020]. Pairwise distances were
calculated by p-distance treating gaps as pairwise deletions. Distance matrices
and FASTA (*.fas*) files with combined headers for duplicate DNA sequences were
deposited in [data-raw/Bayes/MEGA](`r links$data$mega`). Non-identical single locus
(rITS, *rps*, *ycf1*) and concatenated multi-locus FASTA files were used as
input for jModelTest [@Posada2008] to assess evolutionary model fit.
Lastly, alignment gaps were coded using 2matrix [@Salinas2014], a Perl script
that implements “simple indel coding” as described by @Simmons2000.
Partitioned NEXUS (*.nex*) files with defined command blocks were deposited in
[data-raw/Bayes/Runs](`r links$data$runs`).

```{r diagMEGA}
DiagrammeR::mermaid('
graph TD
  S1["sequencing/3.alignments-single/"]
  S2["sequencing/3.multi-locus/"]
  %% MEGA files
  S1-->D1{MEGA}
  S2-->D1
  D1-->D2(Distance matrices)
  D1-->D3(De-duplicated multi-FASTAs)
  D2-->D4["Bayes/MEGA/"]
  D3-->D4
  %% jModelTest
  D4-->J1{jModelTest}
  J1-->J2["Bayes/jmodeltest/"]
  %% 2Matrix / Bayes Nexus
  D4-->M1{2Matrix}
  M1-->M2["Bayes/Runs/"]
  M2-->M4{MrBayes}
  J2-->M4
')
```

## BEAST

The `Biostrings` and `ape` [@R-ape] packages were used to read in the subset of
specimen alignments from the concatenated analysis and write `NEXUS` formatted
alignment files. For all three loci (rITS, *rps*, *ycf1*), alignments were
imported into `BEAUti` v2.6.3 [@Bouckaert2019] to configure BEAST *.xml* files.

```{r BEAUTi}
DiagrammeR::mermaid('
graph TD
  S1["sequencing/3.alignments-subset/"]
  S1-->B1(data-raw/BEAST/beast.R)
  B1-->B2["BEAST/NEXUS/"]
  B2-->B3(BEAUTi)
  B3-->B4(data-raw/BEAST/*.xml)
')
```

BEAST v2.6.3 [@Bouckaert2019] results from 3 independent runs of 50M states were
combined with `LogCombiner` v2.6.3 (*ibid.*) with 10,000 resampled states.
A maximum clade credibility tree was summarized using `TreeAnnotator` v2.6.3
(*ibid.*) and visualized using the `ggplot2` [@R-ggplot2] extension `ggtree`
[@R-ggtree].

```{r BEAST}
DiagrammeR::mermaid('
graph TD
  B1(data-raw/BEAST/*.xml)-->B2
  subgraph ""
    B2[BEAST2]
    B2-->B3[LogCombiner]
    B3-->B4[TreeAnnotator]
  end
  B4-->B5(/inst/extdata/BEAST/*.mcc)
  B5-->B6[R `ggtree`]
')
```

## Evolutionary Models

- https://github.com/ddarriba/modeltest/wiki/Models-of-Evolution

| Locus | Partition | jModelTest | MEGA | Par. Finder | MrBayes | BEAST |
|:------|:----------|:-----------|:-----|:------------|:--------|:------|
| rITS  | 1-659     | TrNef      | HKY  | K80         | nst=1   | TN93  |
| rps   | 660-1605  | F81        | T92  | F81 + i     | nst=1   | JC69  |
| ycf1  | 1606-2162 | F81        | T92  | F81 + i     | nst=1   | JC69  |

## FASTA Summaries

Define a function to read *.fasta* extension files from a given file path as
`DNAStringSet` objects, then calculate the number of unique sequence headers
and minimum / maximum sequence lengths.

[//]: # (TODO <3/27/23>: Optimize FASTA summary function | function export)

```{r fastaSummary}
#| echo: true
# Print aligned FASTA file, number of sequences, and range of alignments
fasta_summaries <- function(fasta_path) {

  # Read FASTA files from path.
  fasta_list <-
    list.files(fasta_path, pattern = "\\.fasta$", full.names = TRUE) %>%
    purrr::map(., Biostrings::readDNAStringSet)

  # Extract locus headers from DNAStringSet.
  fasta_loci <- purrr::map(fasta_list, names) %>%
    purrr::map(., function(locus_headers) {
      purrr::map_chr(locus_headers, function(header) {
        stringr::str_extract(string = header, pattern = "(?<= ).+$")
      }) %>% unique()
    }) %>%
    unlist()

  # Return summary tibble from DNAStringSet object.
  tibble::tibble(
    Locus = unlist(fasta_loci),
    Sequences = purrr::map_int(fasta_list, length),
    `Min Length` = purrr::map_int(
      fasta_list,
      function(fastas) min(Biostrings::width(fastas))
    ),
    `Max Length` = purrr::map_int(
      fasta_list,
      function(fastas) max(Biostrings::width(fastas))
    )
  ) %>%
    knitr::kable(format = "html")
}
```

### Raw FASTAs

```{r summarizeRaw, echo=FALSE}
fasta_summaries(system.file("extdata/FASTA", package = "thesis"))
```

### Subset FASTAs

```{r summarizeSubsets, echo=FALSE}
fasta_summaries("sequencing/2.subset-fastas/")
```

### Aligned (untrimmed)

```{r summarizeUntrimmed, echo=FALSE}
fasta_summaries("sequencing/3.alignments-single/")
```

### Aligned (trimmed)

```{r summarizeTrimmed, echo=FALSE}
fasta_summaries("sequencing/3.alignments-subset/")
```

# References
