# RESULTS {#RESULTS}

\vspace*{-18pt}

```{r sectionBayesianInference}

Thesis::knitr_section(
  knitr_title = "Phylogenetic Analysis",
  knitr_type = "section", chunk_type = knit_variables$pandoc
)

```

```{r phylogeneticsData}

# Function to abbreviate specimen annotations.
abbreviate_spp <- function(label_vector, pandoc_conversion) {

  paste_html <- function(label) {
    paste0("*", label, "*")
  }

  paste_latex <- function(label) {
    paste0("\\textit{", label, "}")
  }
  
  purrr::map_chr(.x = label_vector, .f = function(label) {
    split_label <- gsub("^Physaria", "P.", label) %>%
      strsplit(x = ., split = " ") %>% unlist(x = .)
    if (length(split_label) %in% c(1, 2)) {
      collapsed_label <- paste(split_label, collapse = " ")
      ifelse(
        pandoc_conversion != "latex",
        paste_html(collapsed_label),
        paste_latex(collapsed_label)
      )
    } else if (length(split_label) > 2) {
      collapsed_label <- paste(split_label[1:2], collapse = " ")
      collapsed_label <-
        paste(
          ifelse(
            pandoc_conversion != "latex",
            paste_html(collapsed_label),
            paste_latex(collapsed_label)
          ), "\n", split_label[3],
          ifelse(
            pandoc_conversion != "latex",
            paste_html(split_label[4]),
            paste_latex(split_label[4])
          ), collapse = " "
        )
      ifelse(pandoc_conversion == "latex",
             kableExtra::linebreak(collapsed_label),
             collapsed_label)
    }
  })
}

# Conservered Seqeuences ----

# Multi-taxa rITS node groups
conserved_rITS <-
  list.files(system.file("extdata/MrBayes", package = "Thesis"),
             full.names = TRUE, pattern = "rITS-infile.nex.con.tre") %>%
  conserved_vouchers(tree_file = .,
                     id_column = "prior_id", id_name = "Priors") %>%
  dplyr::mutate(
    Priors = abbreviate_spp(
      label_vector = .data$Priors,
      pandoc_conversion = knit_variables$pandoc
    )
  ) %>%
  dplyr::arrange(`Node Group`, Priors, State)

# Multi-taxa rps intron node groups
conserved_rps <-
  list.files(system.file("extdata/MrBayes", package = "Thesis"),
             full.names = TRUE, pattern = "rps-infile.nex.con.tre") %>%
  conserved_vouchers(tree_file = .,
                     id_column = "prior_id", id_name = "Priors") %>%
  dplyr::mutate(
    Priors = abbreviate_spp(
      label_vector = .data$Priors,
      pandoc_conversion = knit_variables$pandoc
    )
  ) %>%
  dplyr::arrange(`Node Group`, Priors, State)

# Multi-taxa concatenated node groups
conserved_multiLocus <-
  list.files(system.file("extdata/MrBayes", package = "Thesis"),
             full.names = TRUE, pattern = "ml-infile.nex.con.tre") %>%
  conserved_vouchers(tree_file = .,
                     id_column = "prior_id", id_name = "Priors") %>%
  dplyr::mutate(
    Priors = abbreviate_spp(
      label_vector = .data$Priors,
      pandoc_conversion = knit_variables$pandoc
    )
  ) %>%
  dplyr::arrange(`Node Group`, Priors, State)

```

\paragraph{} Phylogenetic analyses with MrBayes were completed for two
individual loci, rITS and *rps* intron. Pairwise distances of the *ycf1* loci
revealed a high degree of sequence similarity with 36 of 51 specimens having
identical genotypes. Individual gene tree analyses also indicated a notable
degree of sequence homogeneity, with seven groupings of identical genotypes for
rITS and nine such groupings for the *rps* intron. In addition to individual
gene trees, a concatenated analysis with sequence data from rITS, *rps* intron,
and *ycf1* yielded an alignment with two groupings of identical genotypes.
Additional phylogenies estimated using BEAST support close relationships among
the southwestern Wyoming endemics and a *Physaria didymocarpa* species complex.

```{r subsectionrITS}

Thesis::knitr_section(
  knitr_title = "ITS Sequence Analysis",
  knitr_type = "subsection", chunk_type = knit_variables$pandoc
)

```

\paragraph{} Sequences for the rITS locus ranged from 581 to 583 base pairs with
an alignment length of 586 base pairs. Of the identical genotype groupings,
four include interspecific specimens when considered by prior annotation
(Table \@ref(tab:TableIdenticalSeqrITS)).
The largest group of 20 specimens (**Genotype 1**) is comprised of southwestern
Wyoming endemics *P. dornii* (n=4), *P. condensata* (n=4), and
*P. integrifolia* (n=3); 6 individuals of *P. acutifolia* from 
southern Wyoming and western Colorado; and three specimens from the Big Horn
Mountains (*Nelson 74315*, *O'Dea 32*, *Ratcliff and O'Kane 48*).
**Genotype 2** includes three specimens of *P. vitulifera* from central Colorado
and one of *P. acutifolia* from central Wyoming. **Genotype 3** includes
three specimens of putative *P. didymocarpa* subsp. *lanata* from Johnson
County, Wyoming and one of *P. brassicoides* from Powder River County, Montana.
Additional groupings include two specimens of *P. didymocarpa* subsp.
*didymocarpa* and *P. saximontana* subsp. *saximontana* from central Wyoming
(**Genotype 4**); two specimens from Idaho previously annotated as
*P. didymocarpa* subsp. *lyrata* (**Genotype 5**); and two sets of two
*P. eburniflora* specimens with identical genotypes (**Genotypes 6** and **7**).
As previously described, no species of were recovered as
monophyletic groups in the rITS gene tree Figure \@ref(fig:FigureBayesrITS).
*P. obcordata*, *P. fendlerii*, and *P. argyraea* were strongly supported as
outgroup taxa (P=1.000). Overall, the rITS gene tree is not well resolved with
a 19 node polytomy including 6 identical genotype groups (1, 2, 3, 5, 6, 7).
The polytomy include specimens previously annotated as *P. vitulifera*,
*P. saximontana* subsp. *dentata*, *P. eburnifora*,
*P. didymocarpa* subsp. *didymocarpa*, *P. d.* subsp. *lanata*,
*P. d.* subsp. *lyrata*, *P. dornii*, *P. condensata*, *P. brassicoides* and
*P. acutifolia*. 

\clearpage

```{r TableIdenticalSeqrITS}

caption$TableIdenticalSeqrITS <-
  paste(
    physaria(knit_variables$pandoc),
    "DNA specimens with identical genotypes for the rITS locus."
  )

if (knit_variables$pandoc != "docx") {
  bayes_kable(
    conserved_specimens = conserved_rITS,
    knitr_chunk = knit_variables$kable_format,
    kable_caption = caption$TableIdenticalSeqrITS
  )
} else {
  kable(
    x = conserved_rITS,
    format = knit_variables$kable_format,
    caption = caption$TableIdenticalSeqrITS
  )
}

```

\clearpage

```{r FigureBayesrITS, fig.cap=caption$FigureBayesrITS, fig.scap="rITS MrBayes gene tree"}

caption$FigureBayesrITS <-
  "MrBayes consensus tree of rITS locus. "

knitr::include_graphics(
  path = fs::path(
    here::here(), "inst/figures/ResultsBayes_rITS",
    ext = knit_variables$graphics
  )
)

```

\paragraph{} In the polytomy, **Genotype 1** includes all sampled
individuals previously annotated as *Physaria condensata*
(*Ratcliff and O'Kane, Jr. 40* and *41*; *O'Kane, Jr. 3787*; *Fertig 16765*) and
*P. dornii* (*Ratcliff and O'Kane, Jr. 42*; *Refsdal and Refsdal 4376*;
*Fertig 17503* and *17506*) in addition to three *P. integrifolia* specimens
(*Nelson 19288*, *Nelson and Scott 68918*, and *Fertig 17493*). An identical
genotype was observed in six specimens of *P. acutifolia* from southern Wyoming
and western Colorado in addition to three specimens from near Bighorn Canyon.
A moderately supported polytomy (P=0.741) includes four specimens
of *P. didymocarpa* subsp. *didymocarpa* (**Genotype 4** in part), two specimens
*P. s.* subsp. *saximontana* (**Genotype 4** in part), three specimens of
*P. saximontana* subsp. *dentata*, four specimens of *P. brassicoides*,
and single specimens of *P. didymocarpa* subsp. *lanata*,
*P. d.* subsp. *lyrata*, *P. acutifolia* from Utah 
(*Holmgren and Holmgren 14222*), and putative *P. vitulifera* from Carbon County
(*Ratcliff and O'Kane, Jr. 34*). An additional two clades recovered
with high support included *P. saximontana* subsp. *saximontana* and
*P. eburniflora* (P=0.995), and *P. integrifolia* (P=0.982).

```{r subsectionrps}

Thesis::knitr_section(
  knitr_title = "*rps* Intron Sequence Analysis",
  knitr_type = "subsection", chunk_type = knit_variables$pandoc
)

```

\paragraph{} Sequences for the *rps* intron ranged from 802 to 852 base pairs
with an aligned length of 889. A high degree of sequence homogeneity was
observed in the *rps* intron alignment (Table \@ref(tab:TableIdenticalSeqrps)).
Fourteen identical sequences (**Genotype 1**) include specimens previously
identified as *P. didymocarpa* subsp. *didymocarpa* (n=2),
*P. d.* subsp. *lanata* (n=2), *P. saximontana* susp. *saximontana* (n=2),
*P. s.* subsp. *dentata* (n=2), *P. acutifolia* (n=3), and *P. brassicoides*
(n=3). These samples are primarily from populations of the Big Horn Mountains
and Absaroka Range of Wyoming and Montana, with outliers in the Big Belt
Mountains of Montana, northeastern Wyoming, and south-central Wyoming.
Another grouping of identical sequences consists of specimens from
western Montana and east-central Idaho (**Genotype 2**), with previous
annotations as *P. didymocarpa* subsp. *didymocarpa* (n=2),
*P. d.* subsp. *lyrata* (n=2), *P. saximontana* subsp. *dentata* (n=2),
and *P. acutifolia* (n=1).

\paragraph{} **Genotype 3** includes specimens annotated as *P. saximontana*
subsp. *saximontana*, and *P. eburniflora* Fremont county near the Wind River
Range, three specimens of *P. acutifolia* from Sublette and Sweetwater Counties
of Wyoming and Rio Blanco County of Colorado, and one specimen (*Dorn 10105*)
from the Medicine Bow Range with a previous annotation as *P. vitulifera*.
A fourth group (**Genotype 4**) includes two other specimens
(*Ratcliff and O'Kane 34*, *Fertig and Welp 19075*) from the Medicine Bow Range
in addition to *P. integrifolia* from Lincoln County, Wyoming; *P. acutifolia*
from Grand County, Utah; and *P. brassicoides* from Converse County, Wyoming.
Five specimens with identical sequences from Lincoln County, Wyoming
(**Genotype 5**) include two collections of *P. dornii*, two of *P. condensata*,
and one collection of *P. integrifolia*.
**Genotype 6** consists of three *P. didymocarpa* subsp. *lanata* specimens
from the Big Horn Mountains and one of *P. acutifolia* from Uinta County,
Wyoming. One specimen of *P. saximontana* subsp. *dentata* from
Glacier County, Montana and one of *P. eburniflora* from Carbon County, Wyoming
form **Genotype 7**. Lastly, two *P. eburniflora* collections from Fremont
County, Wyoming (**Genotype 8**) and two *P. vitulifera* collections from
Colorado (**Genotype 9**) constitute two additional identical genotypes.

```{r TableIdenticalSeqrps}

caption$TableIdenticalSeqrps <-
  paste(
    physaria(knit_variables$pandoc),
    "DNA specimens with identical genotypes for the",
    ifelse(knit_variables$pandoc != "latex",
           "*rps*", "\\textit{rps}"), "intron locus."
  )

if (knit_variables$pandoc == "html") {
  bayes_kable(
    conserved_specimens = conserved_rps,
    knitr_chunk = knit_variables$kable_format,
    kable_caption = caption$TableIdenticalSeqrps
  )
}

if (knit_variables$pandoc == "docx") {
  kable(
    x = conserved_rps,
    format = knit_variables$kable_format,
    caption = caption$TableIdenticalSeqrps
  )
}

if (knit_variables$pandoc == "latex") {
  bayes_kable(
    conserved_specimens = conserved_rps[1:27, ],
    knitr_chunk = knit_variables$kable_format,
    kable_caption = caption$TableIdenticalSeqrps
  )
}

```

<!--- Clear caption formatting to prevent automatic labeling -->
\captionsetup[table]{labelformat=empty}
<!-- <!--- Prevent automatic indexing of table entry in List of Tables -->
\captionsetup[table]{list=no}

```{r TableIdenticalSeqrps2}

if (knit_variables$pandoc == "latex") {
  bayes_kable(
    conserved_specimens = conserved_rps[28:nrow(conserved_rps), ],
    knitr_chunk = knit_variables$kable_format,
    kable_caption = "Table 5. Continued."
  )
}

if (knit_variables$pandoc == "latex") {
  cat("\\captionsetup[table]{labelformat=default}")
  cat("\\captionsetup[table]{list=yes}")
  cat("\\setcounter{table}{5}")
}

```

\paragraph{} Individual gene tree analysis for the *rps* intron did not recover
monophyletic groups of species as previously described
(Figure \@ref(fig:FigureBayesrpsIntron)). *Physaria fendlerii* and *P. argyraea*
are recovered as outgroups while a 13 node polytomy includes *P. vitulifera*
of Colorado and one specimen from the Medicine Bow Range (*Dorn 10105*),
*P. eburniflora*, *P. acutifolia*, *P. saximontana* subsp. *saximontana*,
*P. s.* subsp. *dentata*, *P. didymocarpa* subsp. *didymocarpa*,
*P. d.* subsp. *lyrata*, *P. d.* subsp. *lanata*, *P. integrifolia*, and
*P. obcordata*. Southwestern endemics *P. condensata* (n=3), *P. dornii* (n=3),
and *P. integrifolia* (n=2) were recovered in a clade of moderate support
(P=0.860) with one *P. acutifolia* specimen from Montezuma County, Colorado.
A separate highly supported clade (P=0.980) includes
*P. saximontana* subsp. *saximontana* (n=3), *P. s* subsp. *dentata* (n=2),
*P. didymocarpa* subsp. *didymocarpa* (n=2), *P. d.* subsp. *lanata* (n=2),
*P. brassicoides* (n=1), *P. acutifolia* (n=1) and *P. condensata* (n=1).
Two additional clades include one (P=0.826) with Colorado *P. vitulifera*
specimens (n=2) and another (P=0.996) with *P. brassicoides* (n=1),
*P. integrifolia* (n=1), and three specimens from the Medicine Bow Range
(*Fertig 16713*, *Fertig and Welp 19075*, *Ratcliff and O'Kane, Jr. 34*).

\clearpage

```{r FigureBayesrpsIntron, fig.cap=caption$FigureBayesrpsIntron, fig.scap="\\textit{rps} intron MrBayes gene tree"}

caption$FigureBayesrpsIntron <-
  paste(
    "MrBayes consensus tree of",
  ifelse(knit_variables$pandoc != "latex",
           "*rps*", "\\textit{rps}"),
  "intron gene tree."
  )

knitr::include_graphics(
  path = fs::path(
    here::here(), "inst/figures/ResultsBayes_rps",
    ext = knit_variables$graphics
  )
)

```

\clearpage

\vspace*{-18pt}

```{r subsectionycf1}

Thesis::knitr_section(
  knitr_title = "*ycf1* Sequence Analysis",
  knitr_type = "subsection", chunk_type = knit_variables$pandoc
)

```

\paragraph{} Sequencing of the *ycf1* locus indicated a low mutation rate for
that marker in this group of taxa. Specimens sampled from each of the putative
species considered in this study had identical sequences, with 36 individuals
having an identical genotype (Table \@ref(tab:TableIdenticalSeqycf1)).
Those showing some degree of heterogeneity for this loci included specimens of
*P. acutifolia* (n=3), *P. brassicoides* (n=2), *P. integrifolia* (n=2),
*P. eburniflora* (n=2), *P. condensata* (n=1),
*P. didymocarpa* subsp. *didymocarpa* (n=1), *P. d.* subsp. *lanata* (n=1),
and one specimen from the Medicine Bow Range (*Fertig 16713*). Due to the degree
of sequence homogeneity, a gene tree analysis was not performed.

```{r TableIdenticalSeqycf1}

conserved_ycf1 <-
  list.files(system.file("extdata/MrBayes", package = "Thesis"),
             full.names = TRUE, pattern = "ycf1-identical-seqs.csv") %>%
  readr::read_csv(file = ., col_types = c("dcccd")) %>%
  dplyr::mutate(., 
    Collector = purrr::map_chr(.[["Collector"]], function(collector) {
      gsub("[A-Z]\\. ?", "", collector) %>%
        gsub("&|with", "and", x = .)
      }),
    Species = abbreviate_spp(label_vector = .data$Species, knit_variables$pandoc)
  ) %>%
  dplyr::select(., Genotype, Species, State, Collector, `Collection Number`)

caption$TableIdenticalSeqycf1 <-
  paste(
    physaria(knit_variables$pandoc), 
    "DNA specimens with identical genotypes for the",
    ifelse(knit_variables$pandoc != "latex",
           "*ycf1*", "\\textit{ycf1}"), "loci.")

if (knit_variables$pandoc == "html") {
  bayes_kable(
    conserved_specimens = conserved_ycf1,
    knitr_chunk = knit_variables$kable_format,
    kable_caption = caption$TableIdenticalSeqycf1
  )
}

if (knit_variables$pandoc == "docx") {
  kable(
    x = conserved_ycf1,
    format = knit_variables$kable_format,
    caption = caption$TableIdenticalSeqycf1
  )
}

if (knit_variables$pandoc == "latex") {
    bayes_kable(
      conserved_specimens = conserved_ycf1[1:28, ],
      knitr_chunk = knit_variables$kable_format,
      kable_caption = caption$TableIdenticalSeqycf1
    )
}

```

\clearpage

<!--- Clear caption formatting to prevent automatic labeling -->
\captionsetup[table]{labelformat=empty}
<!-- <!--- Prevent automatic indexing of table entry in List of Tables -->
\captionsetup[table]{list=no}

```{r TableIdenticalSeqycf12}

if (knit_variables$pandoc == "latex") {
  bayes_kable(
    conserved_specimens = conserved_ycf1[29:nrow(conserved_ycf1), ],
    knitr_chunk = knit_variables$kable_format,
    kable_caption = "Table 6. Continued."
  )
}

if (knit_variables$pandoc == "latex") {
  cat("\\captionsetup[table]{labelformat=default}")
  cat("\\captionsetup[table]{list=yes}")
  cat("\\setcounter{table}{6}")
}

```

```{r subsectionMultiLocus}

Thesis::knitr_section(
  knitr_title = "Concatenated Multi-locus Analysis",
  knitr_type = "subsection", chunk_type = knit_variables$pandoc
)

```

\paragraph{} Concatenation of the three sequenced loci yielded an alignment
with more potentially informative characters than any individual gene alignment.
Only two identical genotypes were observed, one with two specimens previously
identified as *P. dornii* and *P. condensata*, and another comprising
of two specimens previously annotated as *P. didymocarpa* subsp. *lyrata*
(Table \@ref(tab:TableIdenticalSeqMultiLocus))
Despite higher overall resolution, 14 nodes were still unresolved as a polytomy
(P=0.777, Figure \@ref(fig:FigureBayesMultiLocus)). Three supported clades
include one with *P. didymocarpa* related species, another with the southwestern
Wyoming endemics, and a third with two Medicine Bow Range specimens
(Figure \@ref(fig:FigureBayesMultiLocus)).

```{r TableIdenticalSeqMultiLocus}

if (knit_variables$pandoc != "docx") {
  bayes_kable(
    conserved_specimens = conserved_multiLocus,
    knitr_chunk = knit_variables$kable_format,
    kable_caption = paste(
      physaria(knit_variables$pandoc),
      "DNA specimens with identical sequences after concatenation."
    )
  )
} else {
  kable(
    x = conserved_multiLocus,
    format = knit_variables$kable_format,
    caption = paste(
      physaria(knit_variables$pandoc),
      "DNA specimens with identical sequences after concatenation."
    )
  )
}

```

\paragraph{} A supported but unresolved clade (P=0.837) includes individuals of
*P. saximontana* subsp. *saximontana* (n=3), *P. s.* subsp. *dentata* (n=2),
*P. didymocarpa* subsp. *lanata* (n=2), *P. brassicoides* (n=3),
*P. acutifolia* (n=2), and one specimen of *P. condensata*. Another clade
(P=0.966) includes individuals annotated as *P. condensata* (n=3), *P. dornii*
(n=2), *P. integrifolia* (n=2), and one *P. acutifolia* individual. 
A third clade (P=0.966) includes two specimens from Carbon County, Wyoming
(*Fertig and Welp 19075*, *Fertig 16713*), *P. acutifolia* (n=1) from Utah,
*P. brassicoides* (n=1) from northeastern Wyoming, and *P. integrifolia* (n=1)
from western Wyoming. Additional supported clades include one (P=0.867) with
*P. vitulifera* (n=2), another (P=1.000) with *P. saximontana* subsp.
*saximontana* (n=1) and *P. eburniflora* (n=1), and a third with poor support
(P=0.504) including *P. didymocarpa* subsp. *lyrata* (n=2) and
*P. brassicoides* (n=1). The unresolved polytomy includes specimens previously
identified as *P. acutifolia* (n=4), *P. didymocarpa* subsp. *didymocarpa* (n=3),
*P. d.* subsp. *lyrata* (n=1), *P. d.* subsp. *lanata* (n=1), and
*P. eburniflora* (n=2), *P. saximontana* subsp. *dentata* (n=1), and
another specimen from Carbon County, Wyoming (*Dorn 10105*).

\clearpage

```{r FigureBayesMultiLocus, fig.cap=caption$FigureBayesMultiLocus, fig.scap="Multilocus concatenated MrBayes consensus tree"}

caption$FigureBayesMultiLocus <-
  "MrBayes consensus tree of concatenated multi-locus data."

knitr::include_graphics(
  path = fs::path(
    here::here(), "inst/figures/ResultsBayes_multiLocus",
    ext = knit_variables$graphics
  )
)

```

\clearpage

```{r FigureResultsBeastMultiLocus, fig.cap=caption$FigureResultsBeastMultiLocus, fig.scap="BEAST multilocus maximum clade credibility tree"}

caption$FigureResultsBeastMultiLocus <-
  "Multilocus partitioned BEAST maximum clade credibility tree."

knitr::include_graphics(
  path = fs::path(
    here::here(), "inst/figures/ResultsBeastMultiLocus",
    ext = knit_variables$graphics
  )
)

```

\clearpage

```{r FigureResultsBeastGeography, fig.cap=caption$FigureResultsBeastGeography, fig.scap="BEAST phylogeography maximum clade credibility tree"}

caption$FigureResultsBeastGeography <-
  "Continuous landscape phylogeography BEAST maximum clade credibility tree."

knitr::include_graphics(
  path = fs::path(
    here::here(), "inst/figures/ResultsBeastGeography",
    ext = knit_variables$graphics
  )
)

```

\clearpage

```{r FigureResultsBeastSpecies, fig.cap=caption$FigureResultsBeastSpecies, fig.scap="*BEAST species trees"}

caption$FigureResultsBeastSpecies <-
  "*BEAST species tree hypotheses."

knitr::include_graphics(
  path = fs::path(
    here::here(), "inst/figures/ResultsBeastSpecies",
    ext = knit_variables$graphics
  )
)

```

