---
title: "Data README"
author: "Ratcliff"
output:
  md_document:
    variant: markdown_github
---

```{r setup, message=FALSE}
# Get full path to project directory with the `here::` package (Müller 2017)
require(here)

# Set the root directory to knit at the project main level
require(knitr)
opts_knit$set(root.dir = here::here())  # Indexed to repository .git file

# Utilizes `magrittr::` %>% pipe operator (Bache and Wickham 2014)
require(magrittr)

# Functions for FASTA reading and DNAStringSet manipulation
require(Biostrings)

# Nifty functional programming tools from tidyverse suite
require(purrr)
```

## Data Workflow

The root directory to knit this *.Rmd* document is set as the project root with the function `here::here()` defined by the top level *.git* file (Müller 2017).  Example paths in this document should be considered from the project root.

# Specimen Data

Records of voucher specimens on loan from lending herbaria (RM, NY, MO, F, ISTC, MONTU, MONT, RSA-POM, UC, UTC, GH, US, CAS, IDS) were compiled into the excel file `Phys_species_totals.xlsx`.  Information recorded from the vouchers included the identification history, collector, collection number, date, institution, locality, geographic coordinates, elevation, ecological description, and measurements of continuous and discrete specimen traits.  The file is located in the `data/1.specimens/` project subdirectory. In total, this project considered 1635 unique collections dated as early as 1844 (Fremont's 2nd Expedition).

```{r specimenData, echo=FALSE, comment=NA}
# List files in specimen data subdirectory excluding uncommitted files.
list.files(path = "data/1.specimens", pattern = "^Phys_species_totals",
           full.names = TRUE) %>% cat(sep = "\n")
```

To read the data into R for downstream phylogenetic, distribution, and morphological analyses, the script `R/specimen_data.R` is sourced within `index.Rmd` during the document build by `bookdown::render_book()` (Xie 2018).  Briefly, the script accomplishes the following:

- Read in data from excel sheets with `specimens_read()`
    - Utilizes `openxlsx::` (Walker 2018)

\bigskip

- Filter records using `dplyr::` (Wickham et al. 2019) to annotations with:
    - *Physaria*
    - *Lesquerella*
    - Brassicaceae

- Check collection date by regular expression with `date_mismatch()`
    - Expected date format is MM/DD/YYYY
    - Write mismatches to the `log/date_logs/` project subdirectory

- Combine data from the *.xlsx* file species sheets into a single data frame
    - Index sheet-respective rows to write *.csv* files after data cleaning

- Parse prior annotations with `parse_priors()`
    - Split the column `$Taxon` containing comma-separated annotation history
    - Account for identification agreement denoted by "!"
    - Combine list into data frame with `plyr::` (Wickham 2011)
    - Extract most recent annotations into a new column

- Update recent annotations by synonym with `parse_synonyms()`
    - Synonyms follow O'Kane (2010)

- Rebuild data frame with parsed annotations and finish cleaning
    - Cast `$Longitude` and `$Latitude` columns as numeric vectors
    - Use `lubridate::` (Grolemund and Wickham 2011) to clean dates
        - Add a column mutation to remove year from date for seasonality

- Write files with `specimens_write()`
    - Sheetname indexed *.csv* files to `output/specimens/`
    - Tables of prior and reviewed identifications to `log/synonyms/`

# Distribution Mapping

TODO - Writeup of `map_tools.R`

## DNA Specimens

The script `R/map_sequenced.R` writes distribution map subsets of specimens with sequencing data.  A data frame for this script is written to `data/dna_map_spp.csv` in the `sequencedHerbariumRecords` chunk of `index.Rmd`.  Each map is scaled to a subset of the taxa and labelled by laboratory accession.

```{r specimenMapSource, eval=FALSE}
# Source script to write PDFs of map sequenced specimens.
source("R/map_sequenced.R")
```

```{r specimenMaps, echo=FALSE, comment=NA}
# List files in the distribution mapping project subdirectory.
list.files("data/2.distributions/sequenced",
           full.names = TRUE) %>% cat(sep = "\n")
```

# References

Bache SM, Wickham H. 2014. magrittr: A forward-pipe operator for R. R package version 1.5. https://CRAN.R-project.org/package=magrittr

Grolemund G, Wickham H. 2011. Dates and times made easy with lubridate. Journal of Statistical Software. 40(3):1-25.

Henry L, Wickham H. 2019. purrr: Functional programming tools. R package version 0.3.0. https://CRAN.R-project.org/package=purrr

Müller K. 2017. here: A simpler way to find your files. R package version 0.1. https://CRAN.R-project.org/package=here

Müller K, Wickham H. 2019. tibble: Simple data frames. R package version 2.0.1. https://CRAN.R-project.org/package=tibble

O’Kane SL. 2010. *Physaria*. In: Flora of North America editorial committee. Flora of North America north of Mexico. Volume 7. Salicaceae to Brassicaceae. New York (NY): Oxford Univ. Press. p. 616-665. 
Payson EB. 1918. Notes on certain Cruciferae. Annals of the Missouri Botanical Garden. 5(2):143-147.

Pagès H, Aboyoun P, Gentleman R, DebRoy S. 2019. Biostrings: Efficient manipulation of biological strings. R package version 2.50.2. https://www.bioconductor.org/packages/release/bioc/html/Biostrings.html 

Walker A. 2018. openxlsx: Read, write and edit XLSX files. R package version 4.1.0. https://CRAN.R-project.org/package=openxlsx

Wickham H. 2011. The split-apply-combine strategy for data analysis. Journal of Statistical Software. 40(1):1-29.

Wickham H, Francois R, Henry L, Müller K. 2019. dplyr: a grammar of data manipulation. R package version 0.8.0.1. https://CRAN.R-project.org/package=dplyr

Xie Y. 2018. bookdown: Authoring books and technical documents with R markdown. R package version 0.9. https://cran.r-project.org/package=bookdown

